<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Genetic Linkage Maps in Autopolyploids • mappoly</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Genetic Linkage Maps in Autopolyploids">
<meta property="og:description" content="Construction of genetic maps in autopolyploid full-sib populations. 
             Uses pairwise recombination fraction estimation as the first 
             source of information to sequentially position allelic variants 
             in specific homologous chromosomes. For situations where pairwise 
             analysis has limited power, the algorithm relies on the multilocus 
             likelihood obtained through a hidden Markov model (HMM). 
             For more detail, please see  Mollinari and Garcia (2019) 
             &lt;doi:10.1534/g3.119.400378&gt; and Mollinari et al. (2020) 
             &lt;doi:10.1534/g3.119.400620&gt;.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">mappoly</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/mappoly_startguide.html">Introduction to MAPpoly</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mmollina/MAPpoly/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<!-- ![](https://raw.githubusercontent.com/mmollina/MAPpoly/master/mappoly_hexsticker.png) --><div class="section level1">
<div class="page-header"><h1 id="mappoly-">MAPpoly <img src="https://raw.githubusercontent.com/mmollina/MAPpoly/main/hex.png" align="right" width="150"><a class="anchor" aria-label="anchor" href="#mappoly-"></a>
</h1></div>
<p>MAPpoly (v. 0.3.2) is an R package to construct genetic maps in autopolyploids with even ploidy levels. In its current version, MAPpoly can handle ploidy levels up to 8 when using hidden Markov models (HMM) and up to 12 when using the two-point simplification. When dealing with large numbers of markers (&gt; 10,000), we strongly recommend using high-performance computing (HPC).</p>
<p><img src="https://raw.githubusercontent.com/mmollina/MAPpoly/master/mappoly.gif"></p>
<p>In its current version, MAPpoly can handle the following types of datasets:</p>
<ol style="list-style-type: decimal">
<li>CSV files</li>
<li>MAPpoly files<ul>
<li>Dosage based</li>
<li>Probability based</li>
</ul>
</li>
<li>
<a href="https://CRAN.R-project.org/package=fitPoly" class="external-link">fitPoly</a> files</li>
<li>VCF files</li>
</ol>
<p>MAPpoly also is capable of importing objects generated by the following R packages</p>
<ol style="list-style-type: decimal">
<li><a href="https://CRAN.R-project.org/package=updog" class="external-link">updog</a></li>
<li><a href="https://CRAN.R-project.org/package=polyRAD" class="external-link">polyRAD</a></li>
<li>
<a href="https://CRAN.R-project.org/package=polymapR" class="external-link">polymapR</a><ul>
<li>Datasets</li>
<li>Maps</li>
</ul>
</li>
</ol>
<p>The mapping strategy uses pairwise recombination fraction estimation as the first source of information to position allelic variants in specific homologs sequentially. The algorithm relies on the multilocus likelihood obtained through a hidden Markov model (HMM) for situations where pairwise analysis has limited power. The derivation of the HMM used in MAPpoly can be found in <a href="https://doi.org/10.1534/g3.119.400378" class="external-link">Mollinari and Garcia, 2019</a>. The computation of the offspring’s genotypes probabilities and haplotype reconstruction, as well as the preferential pairing profiles, is presented in <a href="https://doi.org/10.1534/g3.119.400620" class="external-link">Mollinari et al., 2020</a>.</p>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<div class="section level2">
<h2 id="from-cran-stable-version">From CRAN (stable version)<a class="anchor" aria-label="anchor" href="#from-cran-stable-version"></a>
</h2>
<p>To install MAPpoly from the The Comprehensive R Archive Network (CRAN) use</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"mappoly"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="from-github-development-version">From GitHub (development version)<a class="anchor" aria-label="anchor" href="#from-github-development-version"></a>
</h2>
<p>You can install the development version from Git Hub. Within R, you need to install <code>devtools</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span></code></pre></div>
<p>If you are using Windows, please install the the latest recommended version of <a href="https://cran.r-project.org/bin/windows/Rtools/" class="external-link">Rtools</a>.</p>
<p>To install MAPpoly from Git Hub use</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"mmollina/mappoly"</span>, dependencies<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>For further QTL analysis, we recommend our <a href="https://cran.r-project.org/package=qtlpoly" class="external-link">QTLpoly</a> package. QTLpoly performs random-effect multiple interval mapping (REMIM) in full-sib families of autopolyploid species based on restricted maximum likelihood (REML) estimation and score statistics, as described in <a href="https://doi.org/10.1534/genetics.120.303080" class="external-link">Pereira et al. 2020</a>.</p>
<p>We recently released <a href="https://cran.r-project.org/package=viewpoly" class="external-link">VIEWpoly</a>. VIEWpoly provides a graphical user interface to integrate, visualize and explore results from linkage and quantitative trait loci analysis, together with genomic information for autopolyploid (and diploid) species. The app is meant for interactive use and allows users to optionally upload different sources of information, including gene annotation and alignment files, enabling the exploitation and search for candidate genes in a genome browser. VIEWpoly supports inputs other than MAPpoly’s, including polymapR, diaQTL, QTLpoly, and polyqtlR.</p>
<p><a href="https://www.youtube.com/watch?v=OBt_jebhfeY" class="external-link"><img src="https://img.youtube.com/vi/OBt_jebhfeY/0.jpg" alt="VIEWpoly tutorial"></a></p>
</div>
</div>
<div class="section level1">
<h1 id="mappolys-workflow">MAPpoly’s workflow<a class="anchor" aria-label="anchor" href="#mappolys-workflow"></a>
</h1>
<p><img src="https://raw.githubusercontent.com/mmollina/MAPpoly/main/MAPpoly_workflow.png"></p>
</div>
<div class="section level1">
<h1 id="vignettes">Vignettes<a class="anchor" aria-label="anchor" href="#vignettes"></a>
</h1>
<ul>
<li>To access the MAPpoly vignette from R, use <code>R  vignette("mappoly_startguide")</code>
</li>
<li><a href="https://rpubs.com/mmollin/tetra_mappoly_vignette" class="external-link">Building a genetic map in a tetraploid potato full-sib population using MAPpoly</a></li>
<li><a href="https://mmollina.github.io/tutorials/hexa_fake/haxaploid_map_construction.html" class="external-link">Building a genetic map in an hexaploid full-sib population using MAPpoly</a></li>
<li>Real datasets<ul>
<li><a href="https://github.com/mmollina/MAPpoly_vignettes/tree/master/data/BT" class="external-link">Hexaploid sweetpotato VCF dataset (Beauregard x Tanzania) obtained using VCF2SM</a></li>
<li><a href="https://github.com/mmollina/MAPpoly_vignettes/blob/master/data/SolCAP_dosage" class="external-link">Tetraploid potato with dosage call in MAPpoly format</a></li>
<li><a href="https://github.com/mmollina/MAPpoly_vignettes/blob/master/data/tetra_solcap.csv" class="external-link">Tetraploid potato with dosage call in CSV format</a></li>
<li><a href="https://github.com/mmollina/MAPpoly_vignettes/blob/master/data/SolCAP" class="external-link">Tetraploid potato with dosage probabilities in MAPpoly format</a></li>
<li><a href="https://raw.githubusercontent.com/mmollina/B2721_map/master/cluster_call/B2721_CC.csv" class="external-link">Tetraploid potato in CSV format obtained using ClusterCall</a></li>
<li><a href="https://github.com/mmollina/SCRI/raw/main/data/fitpoly_tetra_call/B2721_scores.zip" class="external-link">Compressed tetraploid potato with dosage probabilities obtained using fitPoly</a></li>
</ul>
</li>
<li>Simulated datasets<ul>
<li><a href="https://github.com/mmollina/MAPpoly_vignettes/blob/master/data/hexafake" class="external-link">Hexaploid simulation with dosage call in MAPpoly format</a></li>
<li><a href="https://github.com/mmollina/MAPpoly_vignettes/blob/master/data/hexafake_geno_dist" class="external-link">Hexaploid simulation with dosage probabilities in MAPpoly format</a></li>
</ul>
</li>
</ul>
</div>
<div class="section level1">
<h1 id="related-software">Related software<a class="anchor" aria-label="anchor" href="#related-software"></a>
</h1>
<ul><li>
<a href="https://polyploids.r-universe.dev" class="external-link">Polyverse</a> - the polyploid R universe (a <a href="https://lvclark.github.io/" class="external-link">Lindsay Clark</a>’s initiative)</li></ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Enable this universe</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    polyploids <span class="op">=</span> <span class="st">'https://polyploids.r-universe.dev'</span>,</span>
<span>    CRAN <span class="op">=</span> <span class="st">'https://cloud.r-project.org'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install some packages</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'mappoly'</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>Variant Calling<ul><li><a href="https://github.com/bodeolukolu/GBSapp" class="external-link">GBSapp: An automated pipeline for variant calling and filtering.</a></li></ul>
</li>
<li>Simulations<ul><li><a href="https://github.com/PBR/pedigreeSim" class="external-link">PedigreeSim: Simulation of genetic marker data in diploid and polyploid pedigreed populations.</a></li></ul>
</li>
<li>Genotype calling<ul>
<li><a href="https://potatobreeding.cals.wisc.edu/software/" class="external-link">ClusterCall: Automated tetraploid genotype calling by hierarchical clustering</a></li>
<li><a href="https://CRAN.R-project.org/package=fitPoly" class="external-link">fitPoly: Genotype Calling for Bi-Allelic Marker Assays</a></li>
<li><a href="https://CRAN.R-project.org/package=polyRAD" class="external-link">polyRAD: Genotype Calling with Uncertainty from Sequencing Data in Polyploids and Diploids</a></li>
<li><a href="https://bitbucket.org/orserang/supermassa" class="external-link">SuperMASSA: Graphical Bayesian inference tool for genotyping polyploids</a></li>
<li><a href="https://CRAN.R-project.org/package=updog" class="external-link">updog: Flexible Genotyping for Polyploids</a></li>
<li><a href="https://github.com/guilherme-pereira/vcf2sm" class="external-link">VCF2SM: Python script that integrates VCF files and SuperMASSA</a></li>
</ul>
</li>
<li>Genetic mapping in polyploids<ul>
<li><a href="https://CRAN.R-project.org/package=MDSMap" class="external-link">MDSMap: High Density Genetic Linkage Mapping using Multidimensional Scaling</a></li>
<li><a href="https://CRAN.R-project.org/package=polymapR" class="external-link">polymapR: Linkage Analysis in Outcrossing Polyploids</a></li>
<li><a href="https://www.bioss.ac.uk/knowledge/tetraploidmap/" class="external-link">TetraploidSNPMap: Linkage maps and mapping QTLs for autotetraploid species, using SNP dosage data.</a></li>
</ul>
</li>
<li>Haplotype reconstruction<ul>
<li><a href="https://github.com/PlantandFoodResearch/MCHap" class="external-link">MCHap: Polyploid micro-haplotype assembly using Markov chain Monte Carlo simulation.</a></li>
<li><a href="https://github.com/chaozhi/TetraOrigin" class="external-link">TetraOrigin:haplotype reconstruction in a full-sib tetraploid family</a></li>
<li><a href="https://github.com/chaozhi/PolyOriginR" class="external-link">PolyOriginR:haplotype reconstruction in polyploid multiparental populations</a></li>
</ul>
</li>
<li>QTL mapping<ul>
<li><a href="https://cran.r-project.org/package=qtlpoly" class="external-link">QTLpoly: QTL mapping in full-sib families of outcrossing autopolyploid species based on a random-effect multiple QTL model</a></li>
<li><a href="https://github.com/jendelman/diaQTL" class="external-link">diaQTL: QTL analysis of diploid and autotetraploid diallel populations</a></li>
<li><a href="https://cran.r-project.org/package=polyqtlR" class="external-link">polyqtlR: QTL analysis and exploration of meiotic patterns in autopolyploid bi-parental F1 populations.</a></li>
</ul>
</li>
<li>Visualization<ul><li><a href="https://cran.r-project.org/package=viewpoly" class="external-link">VIEWpoly: integrate, visualize and explore results from genetic analysis, together with genomic information for autopolyploids</a></li></ul>
</li>
</ul>
</div>
<div class="section level1">
<h1 id="miscellaneous">Miscellaneous<a class="anchor" aria-label="anchor" href="#miscellaneous"></a>
</h1>
<ul>
<li><a href="https://github.com/mmollina/Autopolyploid_Linkage" class="external-link">Supplementary scripts for Mollinari and Garcia (2019)</a></li>
<li><a href="https://github.com/mmollina/MAPpoly_vignettes/blob/master/README.md" class="external-link">Miscellaneous scripts</a></li>
</ul>
</div>
<div class="section level1">
<h1 id="articles-referencing-mappoly">Articles referencing MAPpoly<a class="anchor" aria-label="anchor" href="#articles-referencing-mappoly"></a>
</h1>
<ol style="list-style-type: decimal">
<li>Rose Rosette Disease Resistance Loci Detected in Two Interconnected Tetraploid Garden Rose Populations (<a href="https://doi.org/10.3389%2Ffpls.2022.916231" class="external-link">Lau et al., 2022</a>)</li>
<li>VIEWpoly: a visualization tool to integrate and explore results of polyploid genetic analysis. (<a href="https://doi.org/10.21105/joss.04242" class="external-link">Taniguti et al., 2022</a>)</li>
<li>Genetic Dissection of Early Blight Resistance in Tetraploid Potato. (<a href="https://doi.org/10.3389/fpls.2022.851538" class="external-link">Xue et al., 2022</a>)</li>
<li>Haplotype reconstruction in connected tetraploid F1 populations (<a href="https://doi.org/10.1093/genetics/iyab106" class="external-link">Zheng et al., 2021</a>)</li>
<li>QTL mapping in outbred tetraploid (and diploid) diallel populations (<a href="https://doi.org/10.1093/genetics/iyab124" class="external-link">Amadeu et al., 2021</a>)</li>
<li>Using probabilistic genotypes in linkage analysis of polyploids. (<a href="https://doi.org/10.1007/s00122-021-03834-x" class="external-link">Liao et al., 2021</a>)</li>
<li>Discovery of a major QTL for root-knot nematode <em>Meloidogyne incognita</em> resistance in cultivated sweetpotato <em>Ipomoea batatas</em>. (<a href="https://doi.org/10.1007/s00122-021-03797-z" class="external-link">Oloka, et al., 2021</a>)</li>
<li>Quantitative trait locus mapping for common scab resistance in a tetraploid potato full-sib population. (<a href="https://doi.org/10.1094/PDIS-10-20-2270-RE" class="external-link">Pereira et al., 2021</a>)</li>
<li>The recombination landscape and multiple QTL mapping in a Solanum tuberosum cv.’Atlantic’-derived F1 population. (<a href="https://doi.org/10.1101/2020.08.24.265397" class="external-link">Pereira et al., 2021</a>)</li>
<li>High-Resolution Linkage Map and QTL Analyses of Fruit Firmness in Autotetraploid Blueberry (<a href="https://doi.org/10.3389/fpls.2020.562171" class="external-link">Cappai et al., 2020</a>)</li>
<li>When a phenotype is not the genotype: Implications of phenotype misclassification and pedigree errors in genomics-assisted breeding of sweetpotato <em>Ipomoea batatas</em> (L.) Lam.(<a href="https://doi.org/10.1101/747469" class="external-link">Gemenet et al., 2020</a>)</li>
<li>Quantitative trait loci and differential gene expression analyses reveal the genetic basis for negatively associated beta-carotene and starch content in hexaploid sweetpotato [<em>Ipomoea batatas</em> (L.) Lam.] (<a href="https://doi.org/10.1007/s00122-019-03437-7" class="external-link">Gemenet et al., 2020</a>)</li>
<li>Multiple QTL Mapping in Autopolyploids: A Random-Effect Model Approach with Application in a Hexaploid Sweetpotato Full-Sib Population. (<a href="https://doi.org/10.1534/genetics.120.303080" class="external-link">Pereira et al., 2020</a>)</li>
</ol>
</div>
<div class="section level1">
<h1 id="acknowledgment">Acknowledgment<a class="anchor" aria-label="anchor" href="#acknowledgment"></a>
</h1>
<p>This package has been developed as part of the <a href="https://sweetpotatogenomics.cals.ncsu.edu/" class="external-link">Genomic Tools for Sweetpotato Improvement project</a> (GT4SP) and <a href="https://cgspace.cgiar.org/handle/10568/106838" class="external-link">SweetGAINS</a>, both funded by <a href="https://www.gatesfoundation.org/" class="external-link">Bill &amp; Melinda Gates Foundation</a>. Its continuous improvement is made possible by <a href="https://www.polyploids.org/" class="external-link">Tools for polyploids</a>, funded by USDA NIFA Specialty Crop Research Initiative Award.</p>
<div class="horizontalgap" style="width:5px"><div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="op">&lt;</span>a id=<span class="st">"NCSU"</span> href=<span class="st">"https://www.ncsu.edu/"</span><span class="op">&gt;</span><span class="er">&lt;</span>img src=<span class="st">"https://brand.ncsu.edu/assets/logos/ncstate-brick-2x2-red.png"</span> width=<span class="st">"150"</span> alt=<span class="st">""</span><span class="op">/</span><span class="er">&gt;&lt;/</span>a<span class="op">&gt;</span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="er">&lt;</span>a id=<span class="st">"BMGF"</span> href=<span class="st">"https://www.gatesfoundation.org/"</span><span class="op">&gt;</span><span class="er">&lt;</span>img src=<span class="st">"https://fsm-alliance.org/wp-content/uploads/gates-logo-bda5cc0866e8e37eccab4ac502b916c1-copy.png"</span> width=<span class="st">"150"</span> alt=<span class="st">""</span><span class="op">/</span><span class="er">&gt;&lt;/</span>a<span class="op">&gt;</span></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="er">&lt;</span>a id=<span class="st">"GT4SP"</span> href=<span class="st">"https://sweetpotatogenomics.cals.ncsu.edu/"</span><span class="op">&gt;</span><span class="er">&lt;</span>img src=<span class="st">"http://www.sweetpotatoknowledge.org/wp-content/uploads/2016/02/GT4SP-logo-e1456736272456.png"</span> width=<span class="st">"70"</span> alt=<span class="st">""</span><span class="op">/</span><span class="er">&gt;&lt;/</span>a<span class="op">&gt;</span></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="er">&lt;</span>a id=<span class="st">"sweetgains"</span> href=<span class="st">"https://cgspace.cgiar.org/handle/10568/106838"</span><span class="op">&gt;</span><span class="er">&lt;</span>img src=<span class="st">"https://cipotato.org/wp-content/uploads/2020/06/SweetGains-sin-fondo-1-350x230.png"</span> width=<span class="st">"150"</span> alt=<span class="st">""</span><span class="op">/</span><span class="er">&gt;&lt;/</span>a<span class="op">&gt;</span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="er">&lt;</span>a id=<span class="st">"PolyploidTools"</span> href=<span class="st">"https://www.polyploids.org/"</span><span class="op">&gt;</span><span class="er">&lt;</span>img src=<span class="st">"https://www.polyploids.org/sites/default/files/inline-images/Project%20Logo-transparent.png"</span> width=<span class="st">"180"</span> alt=<span class="st">""</span><span class="op">/</span><span class="er">&gt;&lt;/</span>a<span class="op">&gt;</span><span class="st">    </span></span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="st"> </span><span class="er">&lt;</span>a id=<span class="st">"USDA-NIFA"</span> href=<span class="st">"https://nifa.usda.gov/"</span><span class="op">&gt;</span><span class="er">&lt;</span>img src=<span class="st">"https://upload.wikimedia.org/wikipedia/commons/0/06/USDA_NIFA_Twitter_Logo.jpg"</span> width=<span class="st">"100"</span> alt=<span class="st">""</span><span class="op">/</span><span class="er">&gt;&lt;/</span>a<span class="op">&gt;</span><span class="st">  </span></span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="er">&lt;</span>span class=<span class="st">"stretch"</span><span class="op">&gt;</span><span class="er">&lt;/</span>span<span class="op">&gt;</span></span></code></pre></div></div>
<hr>
<p><sub>NC State University promotes equal opportunity and prohibits discrimination and harassment based upon one’s age, color, disability, gender identity, genetic information, national origin, race, religion, sex (including pregnancy), sexual orientation and veteran status.</sub></p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=mappoly" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/mmollina/MAPpoly/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/mmollina/MAPpoly/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing mappoly</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Marcelo Mollinari <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-7001-8498" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Gabriel Gesteira <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-4106-7346" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/mmollina/mappoly/actions" class="external-link"><img src="https://github.com/mmollina/MAPpoly/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://ci.appveyor.com/project/mmollina/mappoly" class="external-link"><img src="https://ci.appveyor.com/api/projects/status/github/mmollina/mappoly?branch=master&amp;svg=true" alt="AppVeyor Build Status"></a></li>
<li><a href="https://www.gnu.org/licenses/gpl-3.0" class="external-link"><img src="https://img.shields.io/badge/License-GPL%20v3-blue.svg" alt="License: GPL v3"></a></li>
<li><a href="https://codecov.io/github/mmollina/MAPpoly" class="external-link"><img src="https://codecov.io/github/mmollina/MAPpoly/branch/master/graphs/badge.svg" alt="codecov"></a></li>
<li><a href="https://cran.r-project.org/package=mappoly" class="external-link"><img src="http://www.r-pkg.org/badges/version/mappoly" alt="CRAN_Status_Badge"></a></li>
<li><a href="https://polyploids.r-universe.dev" class="external-link"><img src="https://polyploids.r-universe.dev/badges/mappoly" alt="R-universe PolyVerse Status Badge"></a></li>
<li><a href="https://cran.r-project.org/package=mappoly" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/grand-total/mappoly"></a></li>
<li><a href="https://doi.org/10.1534/g3.119.400378" class="external-link"><img src="https://img.shields.io/badge/DOI-10.1534%2Fg3.119.400378-BLUE"></a></li>
<li><a href="https://doi.org/10.1534/g3.119.400620" class="external-link"><img src="https://img.shields.io/badge/DOI-10.1534%2Fg3.119.400620-BLUE"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Marcelo Mollinari, Gabriel Gesteira.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
