<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Add a single marker to a map — add_marker • mappoly</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Add a single marker to a map — add_marker"><meta property="og:description" content="Creates a new map by adding a marker in a given position in a pre-built map."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mappoly</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/mappoly_startguide.html">Introduction to MAPpoly</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mmollina/MAPpoly/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Add a single marker to a map</h1>
    <small class="dont-index">Source: <a href="https://github.com/mmollina/MAPpoly/blob/HEAD/R/add_drop.R" class="external-link"><code>R/add_drop.R</code></a></small>
    <div class="hidden name"><code>add_marker.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Creates a new map by adding a marker in a given position in a pre-built map.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">add_marker</span><span class="op">(</span></span>
<span>  <span class="va">input.map</span>,</span>
<span>  <span class="va">mrk</span>,</span>
<span>  <span class="va">pos</span>,</span>
<span>  <span class="va">rf.matrix</span>,</span>
<span>  genoprob <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  phase.config <span class="op">=</span> <span class="st">"best"</span>,</span>
<span>  tol <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  extend.tail <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  r.test <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>input.map</dt>
<dd><p>an object of class <code>mappoly.map</code></p></dd>


<dt>mrk</dt>
<dd><p>the name of the marker to be inserted</p></dd>


<dt>pos</dt>
<dd><p>the name of the marker after which the new marker should be added.
One also can inform the numeric position (between markers) were the 
new marker should be added. To insert a marker at the beginning of a 
map, use <code>pos = 0</code></p></dd>


<dt>rf.matrix</dt>
<dd><p>an object of class <code>mappoly.rf.matrix</code> containing the recombination
fractions and the number of homologues sharing alleles between pairwise 
markers on <code>input.map</code>. It is important that <code>shared.alleles = TRUE</code>
in function <code><a href="rf_list_to_matrix.html">rf_list_to_matrix</a></code> when computing <code>rf.matrix</code>.</p></dd>


<dt>genoprob</dt>
<dd><p>an object of class <code>mappoly.genoprob</code> containing the genotype probabilities
for all marker positions on <code>input.map</code></p></dd>


<dt>phase.config</dt>
<dd><p>which phase configuration should be used. "best" (default) 
will choose the maximum likelihood configuration</p></dd>


<dt>tol</dt>
<dd><p>the desired accuracy (default = 10e-04)</p></dd>


<dt>extend.tail</dt>
<dd><p>the length of the chain's tail that should
be used to calculate the likelihood of the map. If <code>NULL</code> (default), 
the function uses all markers positioned.</p></dd>


<dt>r.test</dt>
<dd><p>for internal use only</p></dd>


<dt>verbose</dt>
<dd><p>if <code>TRUE</code> (default), the current progress is shown; if
<code>FALSE</code>, no output is produced</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list of class <code>mappoly.map</code> with two elements:</p>


<p>i) info:  a list containing information about the map, regardless of the linkage phase configuration:</p>
<dl><dt>ploidy</dt>
<dd><p>the ploidy level</p></dd>

<dt>n.mrk</dt>
<dd><p>number of markers</p></dd>

<dt>seq.num</dt>
<dd><p>a vector containing the (ordered) indices of markers in the map, 
               according to the input file</p></dd>

<dt>mrk.names</dt>
<dd><p>the names of markers in the map</p></dd>

<dt>seq.dose.p1</dt>
<dd><p>a vector containing the dosage in parent 1 for all markers in the map</p></dd>

<dt>seq.dose.p2</dt>
<dd><p>a vector containing the dosage in parent 2 for all markers in the map</p></dd>

<dt>chrom</dt>
<dd><p>a vector indicating the sequence (usually chromosome) each marker belongs 
                as informed in the input file. If not available, 
                <code>chrom = NULL</code></p></dd>

<dt>genome.pos</dt>
<dd><p>physical position (usually in megabase) of the markers into the sequence</p></dd>

<dt>seq.ref</dt>
<dd><p>reference base used for each marker (i.e. A, T, C, G). If not available, 
                <code>seq.ref = NULL</code></p></dd>
                 
<dt>seq.alt</dt>
<dd><p>alternative base used for each marker (i.e. A, T, C, G). If not available, 
                <code>seq.ref = NULL</code></p></dd>

<dt>chisq.pval</dt>
<dd><p>a vector containing p-values of the chi-squared test of Mendelian 
                  segregation for all markers in the map</p></dd>
                 
<dt>data.name</dt>
<dd><p>name of the dataset of class <code>mappoly.data</code></p></dd>

<dt>ph.thres</dt>
<dd><p>the LOD threshold used to define the linkage phase configurations to test</p></dd>


</dl><p>ii) a list of maps with possible linkage phase configuration. Each map in the list is also a 
   list containing</p>
<dl><dt>seq.num</dt>
<dd><p>a vector containing the (ordered) indices of markers in the map, 
               according to the input file</p></dd>

<dt>seq.rf</dt>
<dd><p>a vector of size (<code>n.mrk - 1</code>) containing a sequence of recombination 
              fraction between the adjacent markers in the map</p></dd>

<dt>seq.ph</dt>
<dd><p>linkage phase configuration for all markers in both parents</p></dd>

<dt>loglike</dt>
<dd><p>the hmm-based multipoint likelihood</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p><code>add_marker</code> splits the input map into two sub-maps to the left and the 
right of the given position. Using the genotype probabilities, it computes 
the log-likelihood of all possible linkage phases under a two-point threshold
inherited from function <code><a href="rf_list_to_matrix.html">rf_list_to_matrix</a></code>.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Marcelo Mollinari, <a href="mailto:mmollin@ncsu.edu">mmollin@ncsu.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="va">sub.map</span> <span class="op">&lt;-</span> <span class="fu"><a href="get_submap.html">get_submap</a></span><span class="op">(</span><span class="va">maps.hexafake</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, reestimate.rf <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     You selected: reestimate.rf = FALSE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     -----------------------------------------</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     The recombination fractions provided were</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     obtained using the marker positions in the </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     input map; For accurate values, plese </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     reestimate the map using functions 'reest_rf', </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'est_full_hmm_with_global_error' or </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'est_full_hmm_with_prior_prob'</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sub.map</span>, mrk.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="add_marker-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fu"><a href="make_seq_mappoly.html">make_seq_mappoly</a></span><span class="op">(</span><span class="va">hexafake</span>, <span class="va">sub.map</span><span class="op">$</span><span class="va">info</span><span class="op">$</span><span class="va">mrk.names</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tpt</span> <span class="op">&lt;-</span> <span class="fu"><a href="est_pairwise_rf.html">est_pairwise_rf</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO: Going singlemode. Using one CPU for calculation.</span>
<span class="r-in"><span><span class="va">rf.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="rf_list_to_matrix.html">rf_list_to_matrix</a></span><span class="op">(</span>input.twopt <span class="op">=</span> <span class="va">tpt</span>,</span></span>
<span class="r-in"><span>                               thresh.LOD.ph <span class="op">=</span> <span class="fl">3</span>, </span></span>
<span class="r-in"><span>                               thresh.LOD.rf <span class="op">=</span> <span class="fl">3</span>,</span></span>
<span class="r-in"><span>                               shared.alleles <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO: Going singlemode. Using one CPU.</span>
<span class="r-in"><span><span class="co">###### Removing marker "M_1" (first) #######</span></span></span>
<span class="r-in"><span><span class="va">mrk.to.remove</span> <span class="op">&lt;-</span> <span class="st">"M_1"</span></span></span>
<span class="r-in"><span><span class="va">input.map</span> <span class="op">&lt;-</span> <span class="fu"><a href="drop_marker.html">drop_marker</a></span><span class="op">(</span><span class="va">sub.map</span>, <span class="va">mrk.to.remove</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     INFO:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     -----------------------------------------</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     The recombination fractions provided were</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     obtained using the marker positions in the </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     input map; For accurate values, plese </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     reestimate the map using functions 'reest_rf', </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'est_full_hmm_with_global_error' or </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'est_full_hmm_with_prior_prob'</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">input.map</span>, mrk.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="add_marker-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co">## Computing conditional probabilities using the resulting map</span></span></span>
<span class="r-in"><span><span class="va">genoprob</span> <span class="op">&lt;-</span> <span class="fu"><a href="calc_genoprob.html">calc_genoprob</a></span><span class="op">(</span><span class="va">input.map</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Ploidy level: 6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Number of markers: 19</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Number of individuals: 300</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	</span>
<span class="r-in"><span><span class="va">res.add.M_1</span> <span class="op">&lt;-</span> <span class="fu">add_marker</span><span class="op">(</span>input.map <span class="op">=</span> <span class="va">input.map</span>,</span></span>
<span class="r-in"><span>                        mrk <span class="op">=</span> <span class="st">"M_1"</span>,</span></span>
<span class="r-in"><span>                        pos <span class="op">=</span> <span class="fl">0</span>,</span></span>
<span class="r-in"><span>                        rf.matrix <span class="op">=</span> <span class="va">rf.matrix</span>,</span></span>
<span class="r-in"><span>                        genoprob <span class="op">=</span> <span class="va">genoprob</span>,</span></span>
<span class="r-in"><span>                        tol <span class="op">=</span> <span class="fl">10e-4</span><span class="op">)</span>  </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ....</span>
<span class="r-in"><span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">res.add.M_1</span>, mrk.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>                       </span></span>
<span class="r-plt img"><img src="add_marker-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span> <span class="va">best.phase</span> <span class="op">&lt;-</span> <span class="va">res.add.M_1</span><span class="op">$</span><span class="va">maps</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">seq.ph</span></span></span>
<span class="r-in"><span> <span class="va">names.id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">best.phase</span><span class="op">$</span><span class="va">P</span><span class="op">)</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="plot_compare_haplotypes.html">plot_compare_haplotypes</a></span><span class="op">(</span>ploidy <span class="op">=</span> <span class="fl">6</span>,</span></span>
<span class="r-in"><span>                         hom.allele.p1 <span class="op">=</span> <span class="va">best.phase</span><span class="op">$</span><span class="va">P</span><span class="op">[</span><span class="va">names.id</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                         hom.allele.q1 <span class="op">=</span> <span class="va">best.phase</span><span class="op">$</span><span class="va">Q</span><span class="op">[</span><span class="va">names.id</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                         hom.allele.p2 <span class="op">=</span> <span class="va">sub.map</span><span class="op">$</span><span class="va">maps</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">seq.ph</span><span class="op">$</span><span class="va">P</span><span class="op">[</span><span class="va">names.id</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                         hom.allele.q2 <span class="op">=</span> <span class="va">sub.map</span><span class="op">$</span><span class="va">maps</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">seq.ph</span><span class="op">$</span><span class="va">Q</span><span class="op">[</span><span class="va">names.id</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="add_marker-4.png" alt="" width="700" height="433"></span>
<span class="r-in"><span>                         </span></span>
<span class="r-in"><span><span class="co">###### Removing marker "M_10" (middle or last) #######</span></span></span>
<span class="r-in"><span><span class="va">mrk.to.remove</span> <span class="op">&lt;-</span> <span class="st">"M_10"</span></span></span>
<span class="r-in"><span><span class="va">input.map</span> <span class="op">&lt;-</span> <span class="fu"><a href="drop_marker.html">drop_marker</a></span><span class="op">(</span><span class="va">sub.map</span>, <span class="va">mrk.to.remove</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     INFO:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     -----------------------------------------</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     The recombination fractions provided were</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     obtained using the marker positions in the </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     input map; For accurate values, plese </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     reestimate the map using functions 'reest_rf', </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'est_full_hmm_with_global_error' or </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'est_full_hmm_with_prior_prob'</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">input.map</span>, mrk.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="add_marker-5.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># Computing conditional probabilities using the resulting map</span></span></span>
<span class="r-in"><span><span class="va">genoprob</span> <span class="op">&lt;-</span> <span class="fu"><a href="calc_genoprob.html">calc_genoprob</a></span><span class="op">(</span><span class="va">input.map</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Ploidy level: 6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Number of markers: 19</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Number of individuals: 300</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	</span>
<span class="r-in"><span><span class="va">res.add.M_10</span> <span class="op">&lt;-</span> <span class="fu">add_marker</span><span class="op">(</span>input.map <span class="op">=</span> <span class="va">input.map</span>,</span></span>
<span class="r-in"><span>                        mrk <span class="op">=</span> <span class="st">"M_10"</span>,</span></span>
<span class="r-in"><span>                        pos <span class="op">=</span> <span class="st">"M_9"</span>,</span></span>
<span class="r-in"><span>                        rf.matrix <span class="op">=</span> <span class="va">rf.matrix</span>,</span></span>
<span class="r-in"><span>                        genoprob <span class="op">=</span> <span class="va">genoprob</span>,</span></span>
<span class="r-in"><span>                        tol <span class="op">=</span> <span class="fl">10e-4</span><span class="op">)</span>  </span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ............................................</span>
<span class="r-in"><span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">res.add.M_10</span>, mrk.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>                       </span></span>
<span class="r-plt img"><img src="add_marker-6.png" alt="" width="700" height="433"></span>
<span class="r-in"><span> <span class="va">best.phase</span> <span class="op">&lt;-</span> <span class="va">res.add.M_10</span><span class="op">$</span><span class="va">maps</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">seq.ph</span></span></span>
<span class="r-in"><span> <span class="va">names.id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">best.phase</span><span class="op">$</span><span class="va">P</span><span class="op">)</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="plot_compare_haplotypes.html">plot_compare_haplotypes</a></span><span class="op">(</span>ploidy <span class="op">=</span> <span class="fl">6</span>,</span></span>
<span class="r-in"><span>                         hom.allele.p1 <span class="op">=</span> <span class="va">best.phase</span><span class="op">$</span><span class="va">P</span><span class="op">[</span><span class="va">names.id</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                         hom.allele.q1 <span class="op">=</span> <span class="va">best.phase</span><span class="op">$</span><span class="va">Q</span><span class="op">[</span><span class="va">names.id</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                         hom.allele.p2 <span class="op">=</span> <span class="va">sub.map</span><span class="op">$</span><span class="va">maps</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">seq.ph</span><span class="op">$</span><span class="va">P</span><span class="op">[</span><span class="va">names.id</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                         hom.allele.q2 <span class="op">=</span> <span class="va">sub.map</span><span class="op">$</span><span class="va">maps</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">seq.ph</span><span class="op">$</span><span class="va">Q</span><span class="op">[</span><span class="va">names.id</span><span class="op">]</span><span class="op">)</span> </span></span>
<span class="r-plt img"><img src="add_marker-7.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marcelo Mollinari, Gabriel Gesteira, Cristiane Taniguti.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

