<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Frequency of genotypes for two-point recombination fraction estimation — cache_counts_twopt • mappoly</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Frequency of genotypes for two-point recombination fraction estimation — cache_counts_twopt"><meta property="og:description" content="Returns the frequency of each genotype for two-point reduction
of dimensionality. The frequency is calculated for all pairwise
combinations and for all possible linkage phase configurations."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mappoly</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/mappoly_startguide.html">Introduction to MAPpoly</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mmollina/MAPpoly/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Frequency of genotypes for two-point recombination fraction estimation</h1>
    <small class="dont-index">Source: <a href="https://github.com/mmollina/MAPpoly/blob/HEAD/R/cache_twopts.R" class="external-link"><code>R/cache_twopts.R</code></a></small>
    <div class="hidden name"><code>cache_counts_twopt.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Returns the frequency of each genotype for two-point reduction
of dimensionality. The frequency is calculated for all pairwise
combinations and for all possible linkage phase configurations.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">cache_counts_twopt</span><span class="op">(</span></span>
<span>  <span class="va">input.seq</span>,</span>
<span>  cached <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cache.prev <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncpus <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  joint.prob <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>input.seq</dt>
<dd><p>an object of class <code>mappoly.sequence</code></p></dd>


<dt>cached</dt>
<dd><p>If <code>TRUE</code>, access the counts for all
linkage phase configurations in a internal file (default = FALSE)</p></dd>


<dt>cache.prev</dt>
<dd><p>an object of class <code>cache.info</code> containing
pre-computed genotype frequencies, obtained with
<code>cache_counts_twopt</code> (optional, default = NULL)</p></dd>


<dt>ncpus</dt>
<dd><p>Number of parallel processes to spawn (default = 1)</p></dd>


<dt>verbose</dt>
<dd><p>If <code>TRUE</code> (default), print the linkage phase
configurations. If <code>cached = TRUE</code>, nothing is
printed, since all linkage phase configurations will be cached.</p></dd>


<dt>joint.prob</dt>
<dd><p>If <code>FALSE</code> (default), returns the frequency of
genotypes for transition probabilities (conditional
probabilities). If <code>TRUE</code> returns the frequency for joint
probabilities. The latter is especially important to compute the 
Fisher's Information for a pair of markers.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An object of class <code>cache.info</code> which contains one (conditional probabilities)
    or two (both conditional and joint probabilities) lists. Each list
    contains all pairs of dosages between parents for all markers
    in the sequence. The names in each list are of the form 'A-B-C-D', where: A
    represents the dosage in parent 1, marker k; B represents the dosage in parent
    1, marker k+1; C represents the dosage in parent 2, marker k;
    and D represents the dosage in parent 2, marker k+1.  For each
    list, the frequencies were computed for all possible linkage
    phase configurations. The frequencies for each linkage phase
    configuration are distributed in matrices whose names
    represents the number of homologous chromosomes that share
    alleles. The rows on these matrices represents the dosages in markers k
    and k+1 for an individual in the offspring. See Table 3 of
    S3 Appendix in Mollinari and Garcia (2019) for an example.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Mollinari, M., and Garcia, A.  A. F. (2019) Linkage
    analysis and haplotype phasing in experimental autopolyploid
    populations with high ploidy level using hidden Markov
    models, _G3: Genes, Genomes, Genetics_. 
    <a href="https://doi.org/10.1534/g3.119.400378" class="external-link">doi:10.1534/g3.119.400378</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Marcelo Mollinari, <a href="mailto:mmollin@ncsu.edu">mmollin@ncsu.edu</a> with updates by Gabriel Gesteira, <a href="mailto:gdesiqu@ncsu.edu">gdesiqu@ncsu.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="va">all.mrk</span> <span class="op">&lt;-</span> <span class="fu"><a href="make_seq_mappoly.html">make_seq_mappoly</a></span><span class="op">(</span><span class="va">tetra.solcap</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="co">## local computation</span></span></span>
<span class="r-in"><span>    <span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">cache_counts_twopt</span><span class="op">(</span><span class="va">all.mrk</span>, ncpus <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Caching the following dosage combination: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         P.k P.k+1 Q.k Q.k+1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.1    0     0   1     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.2    0     1   1     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.3    0     1   1     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.4    0     1   1     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.5    0     2   1     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.6    0     3   1     2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.7    0     3   1     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.8    0     3   1     4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.9    0     4   1     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.10   1     0   0     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.11   1     0   1     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.12   1     1   0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.13   1     1   0     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.14   1     1   1     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.15   1     1   1     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.16   1     1   1     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.17   1     2   0     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.18   1     2   1     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.19   1     2   3     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.20   1     3   0     2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.21   1     3   0     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.22   1     3   0     4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.23   1     3   1     2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.24   1     3   1     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.25   1     3   1     4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.26   1     4   0     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.27   1     4   1     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.28   3     0   2     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.29   3     0   4     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.30   3     1   2     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.31   3     1   2     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.32   3     1   2     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.33   3     1   3     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.34   3     1   4     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.35   3     1   4     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.36   3     1   4     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.37   3     2   2     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.38   3     2   3     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.39   3     2   4     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.40   3     3   2     2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.41   3     3   2     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.42   3     3   2     4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.43   3     3   4     2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.44   3     3   4     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.45   3     3   4     4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.46   3     4   2     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.47   3     4   4     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.48   4     0   3     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.49   4     1   3     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.50   4     1   3     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.51   4     1   3     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.52   4     2   3     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.53   4     3   3     2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.54   4     3   3     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.55   4     3   3     4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Conf.56   4     4   3     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO: Going singlemode. Using one Core/CPU/PC for calculation.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO: Done with 56 phase configurations</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO: Calculation took: 1.648 seconds</span>
<span class="r-in"><span>    <span class="co">## load from internal file or web-stored counts (especially important for high ploidy levels)</span></span></span>
<span class="r-in"><span>    <span class="va">counts.cached</span> <span class="op">&lt;-</span> <span class="fu">cache_counts_twopt</span><span class="op">(</span><span class="va">all.mrk</span>, cached <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marcelo Mollinari, Gabriel Gesteira, Cristiane Taniguti, Jeekin Lau.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

