<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Compute conditional probabilities of the genotypes using probability distribution of dosages — calc_genoprob_dist • mappoly</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Compute conditional probabilities of the genotypes using probability distribution of dosages — calc_genoprob_dist"><meta property="og:description" content="Conditional genotype probabilities are calculated for each marker
position and each individual given a map. In this function,
the probabilities are not calculated between markers."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mappoly</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/mappoly_startguide.html">Introduction to MAPpoly</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mmollina/MAPpoly/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute conditional probabilities of the genotypes using probability distribution of dosages</h1>
    <small class="dont-index">Source: <a href="https://github.com/mmollina/MAPpoly/blob/HEAD/R/calc_genoprob_dist.R" class="external-link"><code>R/calc_genoprob_dist.R</code></a></small>
    <div class="hidden name"><code>calc_genoprob_dist.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Conditional genotype probabilities are calculated for each marker
position and each individual given a map. In this function,
the probabilities are not calculated between markers.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calc_genoprob_dist</span><span class="op">(</span></span>
<span>  <span class="va">input.map</span>,</span>
<span>  dat.prob <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  phase.config <span class="op">=</span> <span class="st">"best"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>input.map</dt>
<dd><p>An object of class <code>mappoly.map</code></p></dd>


<dt>dat.prob</dt>
<dd><p>an object of class <code>mappoly.data</code> containing the
probability distribution of the genotypes</p></dd>


<dt>phase.config</dt>
<dd><p>which phase configuration should be used. "best" (default) 
will choose the phase configuration with the
maximum likelihood</p></dd>


<dt>verbose</dt>
<dd><p>if <code>TRUE</code> (default), the current progress is shown; if
<code>FALSE</code>, no output is produced</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An object of class 'mappoly.genoprob' which has two elements: a tridimensional
array containing the probabilities of all possible genotypes for each individual
in each marker position; and the marker sequence with it's recombination frequencies</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Mollinari, M., and Garcia, A.  A. F. (2019) Linkage
    analysis and haplotype phasing in experimental autopolyploid
    populations with high ploidy level using hidden Markov
    models, _G3: Genes, Genomes, Genetics_. 
    <a href="https://doi.org/10.1534/g3.119.400378" class="external-link">doi:10.1534/g3.119.400378</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Marcelo Mollinari, <a href="mailto:mmollin@ncsu.edu">mmollin@ncsu.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span> <span class="co">## tetraploid example</span></span></span>
<span class="r-in"><span> <span class="va">probs.t</span> <span class="op">&lt;-</span> <span class="fu">calc_genoprob_dist</span><span class="op">(</span>input.map <span class="op">=</span> <span class="va">solcap.prior.map</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                           dat.prob <span class="op">=</span> <span class="va">tetra.solcap.geno.dist</span>,</span></span>
<span class="r-in"><span>                           verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Ploidy level:4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of individuals:160</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..................................................</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	..........</span>
<span class="r-in"><span> <span class="va">probs.t</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   This is an object of class 'mappoly.genoprob'</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -----------------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   No. genotypic classes:                     36 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   No. positions:                             358 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   No. individuals:                           160 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -----------------------------------------------------</span>
<span class="r-in"><span> <span class="co">## displaying individual 1, 36 genotypic states </span></span></span>
<span class="r-in"><span> <span class="co">## (rows) across linkage group 1 (columns)                          </span></span></span>
<span class="r-in"><span> <span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">probs.t</span><span class="op">$</span><span class="va">probs</span><span class="op">[</span>,,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="calc_genoprob_dist-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marcelo Mollinari, Gabriel Gesteira, Cristiane Taniguti.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

