<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Compute conditional probabilities of the genotype (one informative parent) — calc_genoprob_single_parent • mappoly</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Compute conditional probabilities of the genotype (one informative parent) — calc_genoprob_single_parent"><meta property="og:description" content="Conditional genotype probabilities are calculated for each marker
position and each individual given a map"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mappoly</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/mappoly_startguide.html">Introduction to MAPpoly</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mmollina/MAPpoly/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute conditional probabilities of the genotype (one informative parent)</h1>
    <small class="dont-index">Source: <a href="https://github.com/mmollina/MAPpoly/blob/HEAD/R/calc_genoprob_single_parent.R" class="external-link"><code>R/calc_genoprob_single_parent.R</code></a></small>
    <div class="hidden name"><code>calc_genoprob_single_parent.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Conditional genotype probabilities are calculated for each marker
position and each individual given a map</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calc_genoprob_single_parent</span><span class="op">(</span></span>
<span>  <span class="va">input.map</span>,</span>
<span>  step <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  info.parent <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  uninfo.parent <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  global.err <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  phase.config <span class="op">=</span> <span class="st">"best"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>input.map</dt>
<dd><p>An object of class <code>mappoly.map</code> (with exceptions)</p></dd>


<dt>step</dt>
<dd><p>Maximum distance (in cM) between positions at which
the genotype probabilities are calculated, though for
step = 0, probabilities are calculated only at the
marker locations.</p></dd>


<dt>info.parent</dt>
<dd><p>index for informative parent</p></dd>


<dt>uninfo.parent</dt>
<dd><p>index for uninformative parent</p></dd>


<dt>global.err</dt>
<dd><p>the assumed global error rate (default = 0.0)</p></dd>


<dt>phase.config</dt>
<dd><p>which phase configuration should be used. "best" (default)
will choose the phase configuration associated with the
maximum likelihood</p></dd>


<dt>verbose</dt>
<dd><p>if <code>TRUE</code> (default), current progress is shown; if
<code>FALSE</code>, no output is produced</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An object of class 'mappoly.genoprob' which has two elements: a tridimensional
array containing the probabilities of all possible genotypes for each individual
in each marker position; and the marker sequence with it's recombination frequencies</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Mollinari, M., and Garcia, A.  A. F. (2019) Linkage
    analysis and haplotype phasing in experimental autopolyploid
    populations with high ploidy level using hidden Markov
    models, _G3: Genes, Genomes, Genetics_. 
    <a href="https://doi.org/10.1534/g3.119.400378" class="external-link">doi:10.1534/g3.119.400378</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Marcelo Mollinari, <a href="mailto:mmollin@ncsu.edu">mmollin@ncsu.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span> <span class="co">## tetraploid example</span></span></span>
<span class="r-in"><span> <span class="va">s</span> <span class="op">&lt;-</span> <span class="fu"><a href="make_seq_mappoly.html">make_seq_mappoly</a></span><span class="op">(</span><span class="va">tetra.solcap</span>, <span class="st">'seq12'</span>, info.parent <span class="op">=</span> <span class="st">"p1"</span><span class="op">)</span></span></span>
<span class="r-in"><span> <span class="va">tpt</span> <span class="op">&lt;-</span> <span class="fu"><a href="est_pairwise_rf.html">est_pairwise_rf</a></span><span class="op">(</span><span class="va">s</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO: Going singlemode. Using one CPU for calculation.</span>
<span class="r-in"><span> <span class="va">map</span> <span class="op">&lt;-</span> <span class="fu"><a href="est_rf_hmm_sequential.html">est_rf_hmm_sequential</a></span><span class="op">(</span>input.seq <span class="op">=</span> <span class="va">s</span>,</span></span>
<span class="r-in"><span>                              twopt <span class="op">=</span> <span class="va">tpt</span>,</span></span>
<span class="r-in"><span>                               start.set <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>                               thres.twopt <span class="op">=</span> <span class="fl">10</span>, </span></span>
<span class="r-in"><span>                               thres.hmm <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>                               extend.tail <span class="op">=</span> <span class="fl">4</span>,</span></span>
<span class="r-in"><span>                               info.tail <span class="op">=</span> <span class="cn">TRUE</span>, </span></span>
<span class="r-in"><span>                               sub.map.size.diff.limit <span class="op">=</span> <span class="fl">8</span>, </span></span>
<span class="r-in"><span>                               phase.number.limit <span class="op">=</span> <span class="fl">4</span>,</span></span>
<span class="r-in"><span>                               reestimate.single.ph.configuration <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                               tol <span class="op">=</span> <span class="fl">10e-2</span>,</span></span>
<span class="r-in"><span>                               tol.final <span class="op">=</span> <span class="fl">10e-3</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of markers: 53</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ════════════════════════════════════════════════════════════ Initial sequence ══</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10 markers...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> •    Trying sequence: 1 2 3 4 5 6 7 8 9 10 :</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        4 phase(s): . . . . </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ══════════════════════════════════════════════════ Done with initial sequence ══</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11/11:(20.8%)  3462: 1 ph     (1/1)       -- tail: 8  <span style="color: #BB0000;">•</span>|<span style="color: #BB0000;">•</span>| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12/12:(22.6%)  3479: 2 ph     (1/2)       -- tail: 9  <span style="color: #BB0000;">•••</span>| <span style="color: #00BBBB;">••••</span>       <span style="color: #BB0000;">•</span>|<span style="color: #BB0000;">••</span> <span style="color: #00BBBB;">••••</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13/13:(24.5%)  3480: 1 ph     (1/1)       -- tail: 4  |<span style="color: #BB0000;">•••</span> <span style="color: #00BBBB;">••••</span>                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14/14:(26.4%)  3483: 1 ph     (1/1)       -- tail: 4  |<span style="color: #BB0000;">•••</span> <span style="color: #00BBBB;">••••</span>                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15/15:(28.3%)  3484: 1 ph     (1/1)       -- tail: 4  <span style="color: #BB0000;">•</span>||| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 16/16:(30.2%)  3488: 1 ph     (1/1)       -- tail: 5  <span style="color: #BB0000;">•</span>|<span style="color: #BB0000;">••</span> <span style="color: #00BBBB;">••••</span>                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 17/17:(32.1%)  3489: 1 ph     (1/1)       -- tail: 6  |<span style="color: #BB0000;">•••</span> <span style="color: #00BBBB;">••••</span>                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 18/18:(34%)    3490: 1 ph     (1/1)       -- tail: 7  |<span style="color: #BB0000;">•••</span> <span style="color: #00BBBB;">••••</span>                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 19/19:(35.8%)  3491: 1 ph     (1/1)       -- tail: 8  |<span style="color: #BB0000;">•</span>|| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 20/20:(37.7%)  3494: 1 ph     (1/1)       -- tail: 9  |<span style="color: #BB0000;">•••</span> <span style="color: #00BBBB;">••••</span>                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 21/21:(39.6%)  3498: 1 ph     (1/1)       -- tail: 10 <span style="color: #BB0000;">•</span>|<span style="color: #BB0000;">••</span> <span style="color: #00BBBB;">••••</span>                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 22/22:(41.5%)  3500: 4 ph     (1/4)       -- tail: 11 <span style="color: #BB0000;">•••</span>| <span style="color: #00BBBB;">••••</span>  ...  <span style="color: #BB0000;">••</span>|<span style="color: #BB0000;">•</span> <span style="color: #00BBBB;">••••</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 23/23:(43.4%)  3502: 1 ph     (1/1)       -- tail: 4  <span style="color: #BB0000;">••</span>|<span style="color: #BB0000;">•</span> <span style="color: #00BBBB;">••••</span>                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 24/24:(45.3%)  3506: 1 ph     (1/1)       -- tail: 4  ||<span style="color: #BB0000;">•</span>| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 25/25:(47.2%)  3507: 1 ph     (1/1)       -- tail: 5  ||<span style="color: #BB0000;">•</span>| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 26/26:(49.1%)  3510: 1 ph     (1/1)       -- tail: 6  <span style="color: #BB0000;">••</span>|<span style="color: #BB0000;">•</span> <span style="color: #00BBBB;">••••</span>                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 27/27:(50.9%)  3511: 1 ph     (1/1)       -- tail: 7  <span style="color: #BB0000;">•</span>||| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 28/28:(52.8%)  3512: 1 ph     (1/1)       -- tail: 7  ||<span style="color: #BB0000;">•</span>| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 29/29:(54.7%)  3517: 2 ph     (1/2)       -- tail: 8  <span style="color: #BB0000;">••</span>|| <span style="color: #00BBBB;">••••</span>       <span style="color: #BB0000;">•</span>||<span style="color: #BB0000;">•</span> <span style="color: #00BBBB;">••••</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 30/30:(56.6%)  3535: 4 ph     (1/4)       -- tail: 4  <span style="color: #BB0000;">•</span>||| ||||  ...  |<span style="color: #BB0000;">•</span>|| |||| </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 31/31:(58.5%)  3539: 1 ph     (1/1)       -- tail: 4  |<span style="color: #BB0000;">•</span>|| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 32/32:(60.4%)  3548: 4 ph     (1/4)       -- tail: 4  <span style="color: #BB0000;">•</span>||| ||||  ...  |<span style="color: #BB0000;">•</span>|| |||| </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00; font-style: italic;">32: not included (~map extension~)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 32/33:(62.3%)  3557: 4 ph     (1/4)       -- tail: 4  <span style="color: #BB0000;">•••</span>| <span style="color: #00BBBB;">••••</span>  ...  <span style="color: #BB0000;">••</span>|<span style="color: #BB0000;">•</span> <span style="color: #00BBBB;">••••</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00; font-style: italic;">33: not included (~map extension~)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 32/34:(64.2%)  3558: 4 ph     (1/4)       -- tail: 4  <span style="color: #BB0000;">•••</span>| <span style="color: #00BBBB;">••••</span>  ...  <span style="color: #BB0000;">••</span>|<span style="color: #BB0000;">•</span> <span style="color: #00BBBB;">••••</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00; font-style: italic;">34: not included (~map extension~)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 32/35:(66%)    3566: 1 ph     (1/1)       -- tail: 4  |<span style="color: #BB0000;">•</span>|| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 33/36:(67.9%)  3569: 1 ph     (1/1)       -- tail: 5  <span style="color: #BB0000;">•</span>|<span style="color: #BB0000;">••</span> <span style="color: #00BBBB;">••••</span>                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 34/37:(69.8%)  3579: 1 ph     (1/1)       -- tail: 6  <span style="color: #BB0000;">•</span>|<span style="color: #BB0000;">••</span> <span style="color: #00BBBB;">••••</span>                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 35/38:(71.7%)  3585: 1 ph     (1/1)       -- tail: 7  |<span style="color: #BB0000;">••</span>| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 36/39:(73.6%)  3592: 2 ph     (1/2)       -- tail: 7  <span style="color: #BB0000;">•••</span>| ||||       |<span style="color: #BB0000;">•••</span> |||| </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 37/40:(75.5%)  3598: 1 ph     (1/1)       -- tail: 4  <span style="color: #BB0000;">•••</span>| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 38/41:(77.4%)  3602: 1 ph     (1/1)       -- tail: 4  |<span style="color: #BB0000;">•</span>|| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 39/42:(79.2%)  3609: 2 ph     (1/2)       -- tail: 4  <span style="color: #BB0000;">••</span>|| ||||       |<span style="color: #BB0000;">•</span>|<span style="color: #BB0000;">•</span> |||| </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 40/43:(81.1%)  3613: 1 ph     (1/1)       -- tail: 4  <span style="color: #BB0000;">••</span>|| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 41/44:(83%)    3614: 2 ph     (1/2)       -- tail: 4  <span style="color: #BB0000;">•••</span>| <span style="color: #00BBBB;">••••</span>       <span style="color: #BB0000;">••</span>|<span style="color: #BB0000;">•</span> <span style="color: #00BBBB;">••••</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 42/45:(84.9%)  3615: 1 ph     (1/1)       -- tail: 4  ||<span style="color: #BB0000;">•</span>| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 43/46:(86.8%)  3616: 1 ph     (1/1)       -- tail: 5  <span style="color: #BB0000;">••</span>|<span style="color: #BB0000;">•</span> <span style="color: #00BBBB;">••••</span>                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 44/47:(88.7%)  3623: 1 ph     (1/1)       -- tail: 6  ||<span style="color: #BB0000;">•</span>| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 45/48:(90.6%)  3624: 1 ph     (1/1)       -- tail: 7  ||<span style="color: #BB0000;">•</span>| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 46/49:(92.5%)  3625: 1 ph     (1/1)       -- tail: 8  <span style="color: #BB0000;">••</span>|<span style="color: #BB0000;">•</span> <span style="color: #00BBBB;">••••</span>                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 47/50:(94.3%)  3626: 1 ph     (1/1)       -- tail: 9  <span style="color: #BB0000;">••</span>|<span style="color: #BB0000;">•</span> <span style="color: #00BBBB;">••••</span>                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 48/51:(96.2%)  3632: 1 ph     (1/1)       -- tail: 10 <span style="color: #BB0000;">•</span>|<span style="color: #BB0000;">•</span>| ||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 49/52:(98.1%)  3642: 2 ph     (1/2)       -- tail: 9  <span style="color: #BB0000;">•••</span>| <span style="color: #00BBBB;">••••</span>       <span style="color: #BB0000;">•</span>|<span style="color: #BB0000;">••</span> <span style="color: #00BBBB;">••••</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 50/53:(100%)   3649: 4 ph     (1/4)       -- tail: 4  <span style="color: #BB0000;">•••</span>| <span style="color: #00BBBB;">••••</span>  ...  <span style="color: #BB0000;">••</span>|<span style="color: #BB0000;">•</span> <span style="color: #00BBBB;">••••</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ══════════════════════════════════ Reestimating final recombination fractions ══</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Markers in the initial sequence: 53</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Mapped markers                  : 50 (94.3%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ════════════════════════════════════════════════════════════════════════════════</span>
<span class="r-in"><span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">map</span><span class="op">)</span>                                     </span></span>
<span class="r-plt img"><img src="calc_genoprob_single_parent-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span> <span class="va">probs</span> <span class="op">&lt;-</span> <span class="fu">calc_genoprob_single_parent</span><span class="op">(</span>input.map <span class="op">=</span> <span class="va">map</span>, </span></span>
<span class="r-in"><span>                                   info.parent <span class="op">=</span> <span class="fl">1</span>, </span></span>
<span class="r-in"><span>                                   uninfo.parent <span class="op">=</span> <span class="fl">2</span>, </span></span>
<span class="r-in"><span>                                   step <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Ploidy level: 4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Number of markers: 126</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Number of individuals: 160</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span> <span class="va">probs</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   This is an object of class 'mappoly.genoprob'</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -----------------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   No. genotypic classes:                     6 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   No. positions:                             126 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   No. individuals:                           160 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -----------------------------------------------------</span>
<span class="r-in"><span> <span class="co">## displaying individual 1, 6 genotypic states</span></span></span>
<span class="r-in"><span> <span class="co">## (rows) across linkage group 1 (columns)                          </span></span></span>
<span class="r-in"><span> <span class="fu"><a href="https://rdrr.io/r/graphics/image.html" class="external-link">image</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">probs</span><span class="op">$</span><span class="va">probs</span><span class="op">[</span>,,<span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="calc_genoprob_single_parent-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marcelo Mollinari, Gabriel Gesteira, Cristiane Taniguti, Jeekin Lau.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

