<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Multipoint analysis using Hidden Markov Models: Sequential phase elimination — est_rf_hmm_sequential • mappoly</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Multipoint analysis using Hidden Markov Models: Sequential phase elimination — est_rf_hmm_sequential"><meta property="og:description" content="Performs the multipoint analysis proposed by Mollinari and
 Garcia (2019) in a sequence of markers removing unlikely phases
using sequential multipoint information."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mappoly</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/mappoly_startguide.html">Introduction to MAPpoly</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mmollina/MAPpoly/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Multipoint analysis using Hidden Markov Models: Sequential phase elimination</h1>
    <small class="dont-index">Source: <a href="https://github.com/mmollina/MAPpoly/blob/HEAD/R/est_map_hmm.R" class="external-link"><code>R/est_map_hmm.R</code></a></small>
    <div class="hidden name"><code>est_rf_hmm_sequential.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Performs the multipoint analysis proposed by <cite>Mollinari and
 Garcia (2019)</cite> in a sequence of markers removing unlikely phases
using sequential multipoint information.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">est_rf_hmm_sequential</span><span class="op">(</span></span>
<span>  <span class="va">input.seq</span>,</span>
<span>  <span class="va">twopt</span>,</span>
<span>  start.set <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  thres.twopt <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  thres.hmm <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  extend.tail <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  phase.number.limit <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  sub.map.size.diff.limit <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  info.tail <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  reestimate.single.ph.configuration <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  tol <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  tol.final <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  detailed.verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  high.prec <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>input.seq</dt>
<dd><p>an object of class <code>mappoly.sequence</code></p></dd>


<dt>twopt</dt>
<dd><p>an object of class <code>mappoly.twopt</code>
containing the two-point information</p></dd>


<dt>start.set</dt>
<dd><p>number of markers to start the phasing procedure (default = 4)</p></dd>


<dt>thres.twopt</dt>
<dd><p>the LOD threshold used to determine if the linkage
phases compared via two-point analysis should be considered 
for the search space reduction (A.K.A. \(\eta\) in 
<cite>Mollinari and Garcia (2019)</cite>, default = 5)</p></dd>


<dt>thres.hmm</dt>
<dd><p>the LOD threshold used to determine if the linkage
phases compared via hmm analysis should be evaluated in the 
next round of marker inclusion (default = 50)</p></dd>


<dt>extend.tail</dt>
<dd><p>the length of the chain's tail that should
be used to calculate the likelihood of the map. If <code>NULL</code> (default), 
the function uses all markers positioned. Even if <code>info.tail = TRUE</code>, 
it uses at least <code>extend.tail</code> as the tail length</p></dd>


<dt>phase.number.limit</dt>
<dd><p>the maximum number of linkage phases of the sub-maps defined 
by arguments <code>info.tail</code> and <code>extend.tail</code>. Default is 20. If the
size exceeds this limit, the marker will not be inserted. If
<code>Inf</code>, then it will insert all markers.</p></dd>


<dt>sub.map.size.diff.limit</dt>
<dd><p>the maximum accepted length
difference between the current and the previous sub-map defined 
by arguments <code>info.tail</code> and <code>extend.tail</code>. If the
size exceeds this limit, the marker will not be inserted. If
<code>NULL</code>(default), then it will insert all markers.</p></dd>


<dt>info.tail</dt>
<dd><p>if <code>TRUE</code> (default), it uses the complete informative tail
of the chain (i.e. number of markers where all homologous 
(\(ploidy x 2\)) can be distinguished) to calculate the map likelihood</p></dd>


<dt>reestimate.single.ph.configuration</dt>
<dd><p>logical. If <code>FALSE</code> (default)
returns a map without re-estimating the map parameters in cases
where there are only one possible linkage phase configuration</p></dd>


<dt>tol</dt>
<dd><p>the desired accuracy during the sequential phase (default = 10e-02)</p></dd>


<dt>tol.final</dt>
<dd><p>the desired accuracy for the final map (default = 10e-04)</p></dd>


<dt>verbose</dt>
<dd><p>If <code>TRUE</code> (default), current progress is shown; if
<code>FALSE</code>, no output is produced</p></dd>


<dt>detailed.verbose</dt>
<dd><p>If <code>TRUE</code>, the expansion of the current 
submap is shown;</p></dd>


<dt>high.prec</dt>
<dd><p>logical. If <code>TRUE</code> uses high precision 
(long double) numbers in the HMM procedure implemented in C++,
which can take a long time to perform (default = FALSE)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list of class <code>mappoly.map</code> with two elements:</p>


<p>i) info:  a list containing information about the map, regardless of the linkage phase configuration:</p>
<dl><dt>ploidy</dt>
<dd><p>the ploidy level</p></dd>

<dt>n.mrk</dt>
<dd><p>number of markers</p></dd>

<dt>seq.num</dt>
<dd><p>a vector containing the (ordered) indices of markers in the map, 
               according to the input file</p></dd>

<dt>mrk.names</dt>
<dd><p>the names of markers in the map</p></dd>

<dt>seq.dose.p1</dt>
<dd><p>a vector containing the dosage in parent 1 for all markers in the map</p></dd>

<dt>seq.dose.p2</dt>
<dd><p>a vector containing the dosage in parent 2 for all markers in the map</p></dd>

<dt>chrom</dt>
<dd><p>a vector indicating the sequence (usually chromosome) each marker belongs 
                as informed in the input file. If not available, 
                <code>chrom = NULL</code></p></dd>

<dt>genome.pos</dt>
<dd><p>physical position (usually in megabase) of the markers into the sequence</p></dd>

<dt>seq.ref</dt>
<dd><p>reference base used for each marker (i.e. A, T, C, G). If not available, 
                <code>seq.ref = NULL</code></p></dd>

<dt>seq.alt</dt>
<dd><p>alternative base used for each marker (i.e. A, T, C, G). If not available, 
                <code>seq.ref = NULL</code></p></dd>

<dt>chisq.pval</dt>
<dd><p>a vector containing p-values of the chi-squared test of Mendelian 
                  segregation for all markers in the map</p></dd>

<dt>data.name</dt>
<dd><p>name of the dataset of class <code>mappoly.data</code></p></dd>

<dt>ph.thres</dt>
<dd><p>the LOD threshold used to define the linkage phase configurations to test</p></dd>


</dl><p>ii) a list of maps with possible linkage phase configuration. Each map in the list is also a 
   list containing</p>
<dl><dt>seq.num</dt>
<dd><p>a vector containing the (ordered) indices of markers in the map, 
               according to the input file</p></dd>

<dt>seq.rf</dt>
<dd><p>a vector of size (<code>n.mrk - 1</code>) containing a sequence of recombination 
              fraction between the adjacent markers in the map</p></dd>

<dt>seq.ph</dt>
<dd><p>linkage phase configuration for all markers in both parents</p></dd>

<dt>loglike</dt>
<dd><p>the hmm-based multipoint likelihood</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>This function sequentially includes markers into a map given an
ordered sequence. It uses two-point information to eliminate
unlikely linkage phase configurations given <code>thres.twopt</code>. The
search is made within a window of size <code>extend.tail</code>. For the
remaining configurations, the HMM-based likelihood is computed and
the ones that pass the HMM threshold (<code>thres.hmm</code>) are eliminated.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Mollinari, M., and Garcia, A.  A. F. (2019) Linkage
    analysis and haplotype phasing in experimental autopolyploid
    populations with high ploidy level using hidden Markov
    models, _G3: Genes, Genomes, Genetics_. 
    <a href="https://doi.org/10.1534/g3.119.400378" class="external-link">doi:10.1534/g3.119.400378</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Marcelo Mollinari, <a href="mailto:mmollin@ncsu.edu">mmollin@ncsu.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span> <span class="co"># \donttest{</span></span></span>
<span class="r-in"><span>    <span class="va">mrk.subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="make_seq_mappoly.html">make_seq_mappoly</a></span><span class="op">(</span><span class="va">hexafake</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">red.mrk</span> <span class="op">&lt;-</span> <span class="fu"><a href="elim_redundant.html">elim_redundant</a></span><span class="op">(</span><span class="va">mrk.subset</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">unique.mrks</span> <span class="op">&lt;-</span> <span class="fu"><a href="make_seq_mappoly.html">make_seq_mappoly</a></span><span class="op">(</span><span class="va">red.mrk</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">subset.pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="est_pairwise_rf.html">est_pairwise_rf</a></span><span class="op">(</span>input.seq <span class="op">=</span> <span class="va">unique.mrks</span>,</span></span>
<span class="r-in"><span>                                  ncpus <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                                  verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO: Going singlemode. Using one CPU for calculation.</span>
<span class="r-in"><span>    <span class="va">subset.map</span> <span class="op">&lt;-</span> <span class="fu">est_rf_hmm_sequential</span><span class="op">(</span>input.seq <span class="op">=</span> <span class="va">unique.mrks</span>,</span></span>
<span class="r-in"><span>                                        thres.twopt <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>                                        thres.hmm <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>                                        extend.tail <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>                                        tol <span class="op">=</span> <span class="fl">0.1</span>,</span></span>
<span class="r-in"><span>                                        tol.final <span class="op">=</span> <span class="fl">10e-3</span>,</span></span>
<span class="r-in"><span>                                        phase.number.limit <span class="op">=</span> <span class="fl">5</span>,</span></span>
<span class="r-in"><span>                                        twopt <span class="op">=</span> <span class="va">subset.pairs</span>,</span></span>
<span class="r-in"><span>                                        verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of markers: 17</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ════════════════════════════════════════════════════════════ Initial sequence ══</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 markers...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> •    Trying sequence: 1 2 3 4 :</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        6 phase(s): . . . . . . </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ══════════════════════════════════════════════════ Done with initial sequence ══</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 /5 :(29.4%)  5 : 1 ph     (1/1)       -- tail: 4  |||<span style="color: #BB0000;">•</span>|| ||||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 /6 :(35.3%)  7 : 4 ph     (1/4)       -- tail: 5  |<span style="color: #BB0000;">••</span>||| ||<span style="color: #00BBBB;">•</span>|||  ...  |<span style="color: #BB0000;">••</span>||| ||||<span style="color: #00BBBB;">•</span>| </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7 /7 :(41.2%)  8 : 1 ph     (1/1)       -- tail: 6  <span style="color: #BB0000;">•</span>||||| <span style="color: #00BBBB;">••</span>||<span style="color: #00BBBB;">••</span>                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8 /8 :(47.1%)  9 : 2 ph     (1/2)       -- tail: 7  |<span style="color: #BB0000;">•</span>|||| ||||||       ||||<span style="color: #BB0000;">•</span>| |||||| </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00; font-style: italic;">9: not included (*linkage phases*)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 9 /10:(58.8%)  12: 1 ph     (1/1)       -- tail: 8  |<span style="color: #BB0000;">•</span>||<span style="color: #BB0000;">•</span>| ||<span style="color: #00BBBB;">•</span>|||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 10/11:(64.7%)  13: 2 ph     (1/2)       -- tail: 9  |<span style="color: #BB0000;">•</span>|<span style="color: #BB0000;">•</span>|| <span style="color: #00BBBB;">•</span>|<span style="color: #00BBBB;">••</span>||       |<span style="color: #BB0000;">•</span>|<span style="color: #BB0000;">•</span>|| ||<span style="color: #00BBBB;">•••</span>| </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11/12:(70.6%)  14: 1 ph     (1/1)       -- tail: 10 ||<span style="color: #BB0000;">•</span>|<span style="color: #BB0000;">•</span>| ||||||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12/13:(76.5%)  15: 4 ph     (1/4)       -- tail: 11 <span style="color: #BB0000;">•</span>||||| ||||||  ...  |<span style="color: #BB0000;">•</span>|||| |||||| </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13/14:(82.4%)  16: 1 ph     (1/1)       -- tail: 12 |||||| |||<span style="color: #00BBBB;">•</span>||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 14/15:(88.2%)  17: 1 ph     (1/1)       -- tail: 13 ||||<span style="color: #BB0000;">•</span>| ||<span style="color: #00BBBB;">•</span>|||                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00; font-style: italic;">16: not included (*linkage phases*)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 15/17:(100%)   20: 1 ph     (1/1)       -- tail: 14 ||||<span style="color: #BB0000;">•</span>| ||<span style="color: #00BBBB;">•</span>|||                     </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ══════════════════════════════════ Reestimating final recombination fractions ══</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Markers in the initial sequence: 17</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Mapped markers                  : 15 (88.2%)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ════════════════════════════════════════════════════════════════════════════════</span>
<span class="r-in"><span>     <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">subset.map</span>, detailed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is an object of class 'mappoly.map'</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Ploidy level:	 6 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. individuals:	 300 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. markers:	 15 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. linkage phases:	 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ---------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Linkage phase configuration:  1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        log-likelihood:	 -781.3506</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        LOD:		 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	      	 a b c d e f         g h i j k l   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  	 M_1  	 o | | | | |         o o | | | |      0.0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  	 M_2  	 | o | | | |         | | | | | |      0.1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  	 M_3  	 | o o | | |         | | o | | |      0.2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  	 M_4  	 | | | o | |         | | | o | |      0.2 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  	 M_5  	 | | | o | |         | | | | | |      0.3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  	 M_7  	 | o | | o |         | | o | | |      0.3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  	 M_8  	 o | | | | |         o o | | o o      0.4 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  	 M_9  	 | | | | o |         | | | | | |      0.4 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  	 M_12 	 | o | | o |         | | o | | |      0.5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  	 M_13 	 | o | o | |         o | o o | |      0.6 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  	 M_14 	 | | o | o |         | | | | | |      0.7 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  	 M_15 	 | | | | | o         | | | | | |      1.0 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  	 M_16 	 | | | | | |         | | | o | |      1.1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  	 M_17 	 | | | | o |         | | o | | |      1.3 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  	 M_20 	 | | | | o |         | | o | | |      1.6 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  </span>
<span class="r-in"><span>     <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">subset.map</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="est_rf_hmm_sequential-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span>     <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">subset.map</span>, left.lim <span class="op">=</span> <span class="fl">0</span>, right.lim <span class="op">=</span> <span class="fl">1</span>, mrk.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="est_rf_hmm_sequential-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span>     <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">subset.map</span>, phase <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="est_rf_hmm_sequential-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span>     </span></span>
<span class="r-in"><span>     <span class="co">## Retrieving simulated linkage phase</span></span></span>
<span class="r-in"><span>     <span class="va">ph.P</span> <span class="op">&lt;-</span> <span class="va">maps.hexafake</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">maps</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">seq.ph</span><span class="op">$</span><span class="va">P</span></span></span>
<span class="r-in"><span>     <span class="va">ph.Q</span> <span class="op">&lt;-</span> <span class="va">maps.hexafake</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">maps</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">seq.ph</span><span class="op">$</span><span class="va">Q</span></span></span>
<span class="r-in"><span>     <span class="co">## Estimated linkage phase</span></span></span>
<span class="r-in"><span>     <span class="va">ph.P.est</span> <span class="op">&lt;-</span> <span class="va">subset.map</span><span class="op">$</span><span class="va">maps</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">seq.ph</span><span class="op">$</span><span class="va">P</span></span></span>
<span class="r-in"><span>     <span class="va">ph.Q.est</span> <span class="op">&lt;-</span> <span class="va">subset.map</span><span class="op">$</span><span class="va">maps</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">seq.ph</span><span class="op">$</span><span class="va">Q</span></span></span>
<span class="r-in"><span>     <span class="fu"><a href="compare_haplotypes.html">compare_haplotypes</a></span><span class="op">(</span>ploidy <span class="op">=</span> <span class="fl">6</span>, h1 <span class="op">=</span> <span class="va">ph.P</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ph.P.est</span><span class="op">)</span><span class="op">]</span>, h2 <span class="op">=</span> <span class="va">ph.P.est</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $is.same.haplo</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $haplo.ord</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1 2 3 4 5 6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span>     <span class="fu"><a href="compare_haplotypes.html">compare_haplotypes</a></span><span class="op">(</span>ploidy <span class="op">=</span> <span class="fl">6</span>, h1 <span class="op">=</span> <span class="va">ph.Q</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ph.Q.est</span><span class="op">)</span><span class="op">]</span>, h2 <span class="op">=</span> <span class="va">ph.Q.est</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $is.same.haplo</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $haplo.ord</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 2 1 3 4 5 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span>   <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marcelo Mollinari, Gabriel Gesteira.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

