<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Pairwise two-point analysis — est_pairwise_rf • mappoly</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Pairwise two-point analysis — est_pairwise_rf"><meta property="og:description" content="Performs the two-point pairwise analysis between all markers in a sequence. 
For each pair, the function estimates the recombination fraction for all 
possible linkage phase configurations and associated LOD Scores."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mappoly</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/mappoly_startguide.html">Introduction to MAPpoly</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mmollina/MAPpoly/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Pairwise two-point analysis</h1>
    <small class="dont-index">Source: <a href="https://github.com/mmollina/MAPpoly/blob/HEAD/R/pairwise_rf.R" class="external-link"><code>R/pairwise_rf.R</code></a></small>
    <div class="hidden name"><code>est_pairwise_rf.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Performs the two-point pairwise analysis between all markers in a sequence. 
For each pair, the function estimates the recombination fraction for all 
possible linkage phase configurations and associated LOD Scores.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">est_pairwise_rf</span><span class="op">(</span></span>
<span>  <span class="va">input.seq</span>,</span>
<span>  count.cache <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  count.matrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncpus <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  mrk.pairs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  n.batches <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  est.type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"disc"</span>, <span class="st">"prob"</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  memory.warning <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  parallelization.type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PSOCK"</span>, <span class="st">"FORK"</span><span class="op">)</span>,</span>
<span>  tol <span class="op">=</span> <span class="va">.Machine</span><span class="op">$</span><span class="va">double.eps</span><span class="op">^</span><span class="fl">0.25</span>,</span>
<span>  ll <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>input.seq</dt>
<dd><p>an object of class <code>mappoly.sequence</code></p></dd>


<dt>count.cache</dt>
<dd><p>an object of class <code>cache.info</code> containing
pre-computed genotype frequencies, obtained with
<code><a href="cache_counts_twopt.html">cache_counts_twopt</a></code>. If <code>NULL</code> (default),
genotype frequencies are internally loaded.</p></dd>


<dt>count.matrix</dt>
<dd><p>similar to <code>count.cache</code>, but in matrix format. 
Mostly for internal use.</p></dd>


<dt>ncpus</dt>
<dd><p>Number of parallel processes (cores) to spawn (default = 1)</p></dd>


<dt>mrk.pairs</dt>
<dd><p>a matrix of dimensions 2*N, containing N
pairs of markers to be analyzed. If <code>NULL</code> (default), all pairs are
considered</p></dd>


<dt>n.batches</dt>
<dd><p>deprecated. Not available on MAPpoly 0.3.0 or higher</p></dd>


<dt>est.type</dt>
<dd><p>Indicates whether to use the discrete ("disc") or the probabilistic ("prob") dosage scoring 
when estimating the two-point recombination fractions.</p></dd>


<dt>verbose</dt>
<dd><p>If <code>TRUE</code> (default), current progress is shown; if
<code>FALSE</code>, no output is produced</p></dd>


<dt>memory.warning</dt>
<dd><p>if <code>TRUE</code>, prints a memory warning if the 
number of markers is greater than 10000 for ploidy levels up to 4, and 
3000 for ploidy levels &gt; 4.</p></dd>


<dt>parallelization.type</dt>
<dd><p>one of the supported cluster types. This should 
be either PSOCK (default) or FORK.</p></dd>


<dt>tol</dt>
<dd><p>the desired accuracy. See <code><a href="https://rdrr.io/r/stats/optimize.html" class="external-link">optimize()</a></code> for details</p></dd>


<dt>ll</dt>
<dd><p>will return log-likelihood instead of LOD scores. 
(for internal use)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An object of class <code>mappoly.twopt</code> which
    is a list  containing the following components:</p>
<dl><dt>data.name</dt>
<dd><p>name of the object of class
        <code>mappoly.data</code> with the raw data</p></dd>

    <dt>n.mrk</dt>
<dd><p>number of markers in the sequence</p></dd>

    <dt>seq.num</dt>
<dd><p>a <code>vector</code> containing the (ordered) indices
        of markers in the sequence, according to the input file</p></dd>

    <dt>pairwise</dt>
<dd><p>a list of size
    <code>choose(length(input.seq$seq.num), 2)</code>, each of them containing a 
    matrix where the name of the rows have the form x-y, where x and y indicate 
    how many homologues share the same allelic variant in parents P and Q, 
    respectively (see Mollinari and Garcia, 2019 for notation). The first 
    column indicates the LOD Score in relation to the most likely linkage 
    phase configuration. The second column shows the estimated recombination 
    fraction for each configuration, and the third indicates the LOD Score 
    comparing the likelihood under no linkage (r = 0.5) with the estimated 
    recombination fraction (evidence of linkage).</p></dd>

    </dl><p><code>chisq.pval.thres</code>threshold used to perform the segregation tests</p>
<p></p>
<p><code>chisq.pval</code>p-values associated with the performed segregation tests</p>


    </div>
    <div id="references">
    <h2>References</h2>
    <p>Mollinari, M., and Garcia, A.  A. F. (2019) Linkage
    analysis and haplotype phasing in experimental autopolyploid
    populations with high ploidy level using hidden Markov
    models, _G3: Genes, Genomes, Genetics_. 
    <a href="https://doi.org/10.1534/g3.119.400378" class="external-link">doi:10.1534/g3.119.400378</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Marcelo Mollinari, <a href="mailto:mmollin@ncsu.edu">mmollin@ncsu.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>  <span class="co">## Tetraploid example (first 50 markers) </span></span></span>
<span class="r-in"><span>  <span class="va">all.mrk</span> <span class="op">&lt;-</span> <span class="fu"><a href="make_seq_mappoly.html">make_seq_mappoly</a></span><span class="op">(</span><span class="va">tetra.solcap</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">red.mrk</span> <span class="op">&lt;-</span> <span class="fu"><a href="elim_redundant.html">elim_redundant</a></span><span class="op">(</span><span class="va">all.mrk</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">unique.mrks</span> <span class="op">&lt;-</span> <span class="fu"><a href="make_seq_mappoly.html">make_seq_mappoly</a></span><span class="op">(</span><span class="va">red.mrk</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">all.pairs</span> <span class="op">&lt;-</span> <span class="fu">est_pairwise_rf</span><span class="op">(</span>input.seq <span class="op">=</span> <span class="va">unique.mrks</span>,</span></span>
<span class="r-in"><span>                               ncpus <span class="op">=</span> <span class="fl">1</span>, </span></span>
<span class="r-in"><span>                               verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO: Going singlemode. Using one CPU for calculation.</span>
<span class="r-in"><span>   <span class="va">all.pairs</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   This is an object of class 'mappoly.twopt'</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -----------------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   No. markers:                             50 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   No. estimated recombination fractions:   940 (76.7%)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -----------------------------------------------------</span>
<span class="r-in"><span>   <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">all.pairs</span>, <span class="fl">20</span>, <span class="fl">21</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="est_pairwise_rf-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span>   <span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="rf_list_to_matrix.html">rf_list_to_matrix</a></span><span class="op">(</span><span class="va">all.pairs</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO: Going singlemode. Using one CPU.</span>
<span class="r-in"><span>   <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="est_pairwise_rf-2.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marcelo Mollinari, Gabriel Gesteira.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

