<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Re-estimate genetic map given a global genotyping error — est_full_hmm_with_global_error • mappoly</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Re-estimate genetic map given a global genotyping error — est_full_hmm_with_global_error"><meta property="og:description" content="This function considers a global error when re-estimating
a genetic map using Hidden Markov models. Since this function 
uses the whole transition space in the HMM, its computation 
can take a while, especially for hexaploid maps."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mappoly</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/mappoly_startguide.html">Introduction to MAPpoly</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mmollina/MAPpoly/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Re-estimate genetic map given a global genotyping error</h1>
    <small class="dont-index">Source: <a href="https://github.com/mmollina/MAPpoly/blob/HEAD/R/reest_map_error.R" class="external-link"><code>R/reest_map_error.R</code></a></small>
    <div class="hidden name"><code>est_full_hmm_with_global_error.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function considers a global error when re-estimating
a genetic map using Hidden Markov models. Since this function 
uses the whole transition space in the HMM, its computation 
can take a while, especially for hexaploid maps.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">est_full_hmm_with_global_error</span><span class="op">(</span></span>
<span>  <span class="va">input.map</span>,</span>
<span>  error <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tol <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  restricted <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  th.prob <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>input.map</dt>
<dd><p>an object of class <code>mappoly.map</code></p></dd>


<dt>error</dt>
<dd><p>the assumed global error rate (default = NULL)</p></dd>


<dt>tol</dt>
<dd><p>the desired accuracy (default = 10e-04)</p></dd>


<dt>restricted</dt>
<dd><p>if <code>TRUE</code> (default), restricts the prior to the 
possible classes under Mendelian, non double-reduced segregation 
given dosage of the parents</p></dd>


<dt>th.prob</dt>
<dd><p>the threshold for using global error or genotype 
probability distribution if present in the dataset (default = 0.95)</p></dd>


<dt>verbose</dt>
<dd><p>if <code>TRUE</code>, current progress is shown; if
<code>FALSE</code> (default), no output is produced</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list of class <code>mappoly.map</code> with two elements:</p>


<p>i) info:  a list containing information about the map, regardless of the linkage phase configuration:</p>
<dl><dt>ploidy</dt>
<dd><p>the ploidy level</p></dd>

<dt>n.mrk</dt>
<dd><p>number of markers</p></dd>

<dt>seq.num</dt>
<dd><p>a vector containing the (ordered) indices of markers in the map, 
               according to the input file</p></dd>

<dt>mrk.names</dt>
<dd><p>the names of markers in the map</p></dd>

<dt>seq.dose.p1</dt>
<dd><p>a vector containing the dosage in parent 1 for all markers in the map</p></dd>

<dt>seq.dose.p2</dt>
<dd><p>a vector containing the dosage in parent 2 for all markers in the map</p></dd>

<dt>chrom</dt>
<dd><p>a vector indicating the sequence (usually chromosome) each marker belongs 
                as informed in the input file. If not available, 
                <code>chrom = NULL</code></p></dd>

<dt>genome.pos</dt>
<dd><p>physical position (usually in megabase) of the markers into the sequence</p></dd>

<dt>seq.ref</dt>
<dd><p>reference base used for each marker (i.e. A, T, C, G). If not available, 
                <code>seq.ref = NULL</code></p></dd>
                 
<dt>seq.alt</dt>
<dd><p>alternative base used for each marker (i.e. A, T, C, G). If not available, 
                <code>seq.ref = NULL</code></p></dd>

<dt>chisq.pval</dt>
<dd><p>a vector containing p-values of the chi-squared test of Mendelian 
                  segregation for all markers in the map</p></dd>
                 
<dt>data.name</dt>
<dd><p>name of the dataset of class <code>mappoly.data</code></p></dd>

<dt>ph.thres</dt>
<dd><p>the LOD threshold used to define the linkage phase configurations to test</p></dd>


</dl><p>ii) a list of maps with possible linkage phase configuration. Each map in the list is also a 
   list containing</p>
<dl><dt>seq.num</dt>
<dd><p>a vector containing the (ordered) indices of markers in the map, 
               according to the input file</p></dd>

<dt>seq.rf</dt>
<dd><p>a vector of size (<code>n.mrk - 1</code>) containing a sequence of recombination 
              fraction between the adjacent markers in the map</p></dd>

<dt>seq.ph</dt>
<dd><p>linkage phase configuration for all markers in both parents</p></dd>

<dt>loglike</dt>
<dd><p>the hmm-based multipoint likelihood</p></dd>

</dl></div>
    <div id="references">
    <h2>References</h2>
    <p>Mollinari, M., and Garcia, A.  A. F. (2019) Linkage
    analysis and haplotype phasing in experimental autopolyploid
    populations with high ploidy level using hidden Markov
    models, _G3: Genes, Genomes, Genetics_. 
    <a href="https://doi.org/10.1534/g3.119.400378" class="external-link">doi:10.1534/g3.119.400378</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Marcelo Mollinari, <a href="mailto:mmollin@ncsu.edu">mmollin@ncsu.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="va">submap</span> <span class="op">&lt;-</span> <span class="fu"><a href="get_submap.html">get_submap</a></span><span class="op">(</span><span class="va">solcap.dose.map</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, mrk.pos <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">err.submap</span> <span class="op">&lt;-</span> <span class="fu">est_full_hmm_with_global_error</span><span class="op">(</span><span class="va">submap</span>, </span></span>
<span class="r-in"><span>                                                 error <span class="op">=</span> <span class="fl">0.01</span>, </span></span>
<span class="r-in"><span>                                                 tol <span class="op">=</span> <span class="fl">10e-4</span>, </span></span>
<span class="r-in"><span>                                                 verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  ----------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  INFO: running HMM using full transition space:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        this operation may take a while.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> -----------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Ploidy level:4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of individuals:160</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.030 0.000 0.000 0.007 0.007 0.005 0.002 0.006 0.003 0.001 0.002 0.000 0.003 0.000 0.003 0.001 0.023 0.037 0.000 	</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.030 0.000 0.000 0.007 0.007 0.003 0.001 0.004 0.001 0.000 0.001 0.000 0.003 0.000 0.001 0.000 0.023 0.037 0.000 	</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.029 0.000 0.000 0.008 0.008 0.003 0.001 0.003 0.000 0.000 0.001 0.000 0.003 0.000 0.001 0.000 0.022 0.037 0.000 	</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.029 0.000 0.000 0.008 0.008 0.003 0.001 0.003 0.000 0.000 0.001 0.000 0.003 0.000 0.001 0.000 0.022 0.037 0.000 	</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.029 0.000 0.000 0.008 0.008 0.003 0.001 0.003 0.000 0.000 0.000 0.000 0.003 0.000 0.000 0.000 0.022 0.037 0.000 	</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.029 0.000 0.000 0.008 0.008 0.003 0.000 0.004 0.000 0.000 0.000 0.000 0.003 0.000 0.000 0.000 0.022 0.037 0.000 	</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.029 0.000 0.000 0.008 0.008 0.003 0.000 0.004 0.000 0.000 0.000 0.000 0.003 0.000 0.000 0.000 0.022 0.037 0.000 </span>
<span class="r-in"><span>    <span class="va">err.submap</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is an object of class 'mappoly.map'</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Ploidy level:	 4 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. individuals:	 160 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. markers:	 20 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. linkage phases:	 1 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ---------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Number of linkage phase configurations:  1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ---------------------------------------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Linkage phase configuration:  1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        map length:	 11.69</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        log-likelihood:	 -377</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        LOD:		 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ~~~~~~~~~~~~~~~~~~</span>
<span class="r-in"><span>    <span class="fu"><a href="plot_map_list.html">plot_map_list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>dose <span class="op">=</span> <span class="va">submap</span>, err <span class="op">=</span> <span class="va">err.submap</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>                  title <span class="op">=</span> <span class="st">"estimation procedure"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="est_full_hmm_with_global_error-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marcelo Mollinari, Gabriel Gesteira.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

