<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ mappoly</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mappoly</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/mappoly_startguide.html">Introduction to MAPpoly</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mmollina/MAPpoly/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="add_marker.html">add_marker()</a></code> </p>
        </td>
        <td><p>Add a single marker to a map</p></td>
      </tr><tr><td>
          <p><code><a href="cache_counts_twopt.html">cache_counts_twopt()</a></code> </p>
        </td>
        <td><p>Frequency of genotypes for two-point recombination fraction estimation</p></td>
      </tr><tr><td>
          <p><code><a href="calc_genoprob.html">calc_genoprob()</a></code> </p>
        </td>
        <td><p>Compute conditional probabilities of the genotypes</p></td>
      </tr><tr><td>
          <p><code><a href="calc_genoprob_dist.html">calc_genoprob_dist()</a></code> </p>
        </td>
        <td><p>Compute conditional probabilities of the genotypes using probability distribution of dosages</p></td>
      </tr><tr><td>
          <p><code><a href="calc_genoprob_error.html">calc_genoprob_error()</a></code> </p>
        </td>
        <td><p>Compute conditional probabilities of the genotypes using global error</p></td>
      </tr><tr><td>
          <p><code><a href="calc_genoprob_single_parent.html">calc_genoprob_single_parent()</a></code> </p>
        </td>
        <td><p>Compute conditional probabilities of the genotype (one informative parent)</p></td>
      </tr><tr><td>
          <p><code><a href="calc_homologprob.html">calc_homologprob()</a></code> </p>
        </td>
        <td><p>Homolog probabilities</p></td>
      </tr><tr><td>
          <p><code><a href="calc_prefpair_profiles.html">calc_prefpair_profiles()</a></code> </p>
        </td>
        <td><p>Preferential pairing profiles</p></td>
      </tr><tr><td>
          <p><code><a href="check_data_sanity.html">check_data_sanity()</a></code> </p>
        </td>
        <td><p>Data sanity check</p></td>
      </tr><tr><td>
          <p><code><a href="compare_maps.html">compare_maps()</a></code> </p>
        </td>
        <td><p>Compare a list of maps</p></td>
      </tr><tr><td>
          <p><code><a href="cross_simulate.html">cross_simulate()</a></code> </p>
        </td>
        <td><p>Simulate an autopolyploid full-sib population</p></td>
      </tr><tr><td>
          <p><code><a href="detect_info_par.html">detect_info_par()</a></code> </p>
        </td>
        <td><p>Detects which parent is informative</p></td>
      </tr><tr><td>
          <p><code><a href="drop_marker.html">drop_marker()</a></code> </p>
        </td>
        <td><p>Remove markers from a map</p></td>
      </tr><tr><td>
          <p><code><a href="edit_order.html">edit_order()</a></code> </p>
        </td>
        <td><p>Edit sequence ordered by reference genome positions
comparing to another set order</p></td>
      </tr><tr><td>
          <p><code><a href="elim_redundant.html">elim_redundant()</a></code> </p>
        </td>
        <td><p>Eliminate redundant markers</p></td>
      </tr><tr><td>
          <p><code><a href="est_full_hmm_with_global_error.html">est_full_hmm_with_global_error()</a></code> </p>
        </td>
        <td><p>Re-estimate genetic map given a global genotyping error</p></td>
      </tr><tr><td>
          <p><code><a href="est_full_hmm_with_prior_prob.html">est_full_hmm_with_prior_prob()</a></code> </p>
        </td>
        <td><p>Re-estimate genetic map using dosage prior probability distribution</p></td>
      </tr><tr><td>
          <p><code><a href="est_pairwise_rf.html">est_pairwise_rf()</a></code> </p>
        </td>
        <td><p>Pairwise two-point analysis</p></td>
      </tr><tr><td>
          <p><code><a href="est_pairwise_rf2.html">est_pairwise_rf2()</a></code> </p>
        </td>
        <td><p>Pairwise two-point analysis - RcppParallel version</p></td>
      </tr><tr><td>
          <p><code><a href="est_rf_hmm.html">est_rf_hmm()</a></code> <code><a href="est_rf_hmm.html">print(<i>&lt;mappoly.map&gt;</i>)</a></code> <code><a href="est_rf_hmm.html">plot(<i>&lt;mappoly.map&gt;</i>)</a></code> </p>
        </td>
        <td><p>Multipoint analysis using Hidden Markov Models in autopolyploids</p></td>
      </tr><tr><td>
          <p><code><a href="est_rf_hmm_sequential.html">est_rf_hmm_sequential()</a></code> </p>
        </td>
        <td><p>Multipoint analysis using Hidden Markov Models: Sequential phase elimination</p></td>
      </tr><tr><td>
          <p><code><a href="export_data_to_polymapR.html">export_data_to_polymapR()</a></code> </p>
        </td>
        <td><p>Export data to <code>polymapR</code></p></td>
      </tr><tr><td>
          <p><code><a href="export_map_list.html">export_map_list()</a></code> </p>
        </td>
        <td><p>Export a genetic map to a CSV file</p></td>
      </tr><tr><td>
          <p><code><a href="export_qtlpoly.html">export_qtlpoly()</a></code> </p>
        </td>
        <td><p>Export to QTLpoly</p></td>
      </tr><tr><td>
          <p><code><a href="extract_map.html">extract_map()</a></code> </p>
        </td>
        <td><p>Extract the maker position from an object of class 'mappoly.map'</p></td>
      </tr><tr><td>
          <p><code><a href="filter_aneuploid.html">filter_aneuploid()</a></code> </p>
        </td>
        <td><p>Filter aneuploid chromosomes from progeny individuals</p></td>
      </tr><tr><td>
          <p><code><a href="filter_individuals.html">filter_individuals()</a></code> </p>
        </td>
        <td><p>Filter out individuals</p></td>
      </tr><tr><td>
          <p><code><a href="filter_missing.html">filter_missing()</a></code> </p>
        </td>
        <td><p>Filter missing genotypes</p></td>
      </tr><tr><td>
          <p><code><a href="filter_segregation.html">filter_segregation()</a></code> </p>
        </td>
        <td><p>Filter markers based on chi-square test</p></td>
      </tr><tr><td>
          <p><code><a href="find_blocks.html">find_blocks()</a></code> </p>
        </td>
        <td><p>Allocate markers into linkage blocks</p></td>
      </tr><tr><td>
          <p><code><a href="framework_map.html">framework_map()</a></code> </p>
        </td>
        <td><p>Design linkage map framework in two steps: i) estimating the recombination fraction with 
HMM approach for each parent separately using only markers segregating individually 
(e.g. map 1 - P1:3 x P2:0, P1: 2x4; map 2 - P1:0 x P2:3, P1:4 x P2:2); ii) merging both 
maps and re-estimate recombination fractions.</p></td>
      </tr><tr><td>
          <p><code><a href="get_genomic_order.html">get_genomic_order()</a></code> <code><a href="get_genomic_order.html">print(<i>&lt;mappoly.geno.ord&gt;</i>)</a></code> <code><a href="get_genomic_order.html">plot(<i>&lt;mappoly.geno.ord&gt;</i>)</a></code> </p>
        </td>
        <td><p>Get the genomic position of markers in a sequence</p></td>
      </tr><tr><td>
          <p><code><a href="get_submap.html">get_submap()</a></code> </p>
        </td>
        <td><p>Extract sub-map from map</p></td>
      </tr><tr><td>
          <p><code><a href="get_tab_mrks.html">get_tab_mrks()</a></code> </p>
        </td>
        <td><p>Get table of dosage combinations</p></td>
      </tr><tr><td>
          <p><code><a href="group_mappoly.html">group_mappoly()</a></code> </p>
        </td>
        <td><p>Assign markers to linkage groups</p></td>
      </tr><tr><td>
          <p><code><a href="hexafake.html">hexafake</a></code> </p>
        </td>
        <td><p>Simulated autohexaploid dataset.</p></td>
      </tr><tr><td>
          <p><code><a href="hexafake.geno.dist.html">hexafake.geno.dist</a></code> </p>
        </td>
        <td><p>Simulated autohexaploid dataset with genotype probabilities.</p></td>
      </tr><tr><td>
          <p><code><a href="import_data_from_polymapR.html">import_data_from_polymapR()</a></code> </p>
        </td>
        <td><p>Import data from polymapR</p></td>
      </tr><tr><td>
          <p><code><a href="import_from_updog.html">import_from_updog()</a></code> </p>
        </td>
        <td><p>Import from updog</p></td>
      </tr><tr><td>
          <p><code><a href="import_phased_maplist_from_polymapR.html">import_phased_maplist_from_polymapR()</a></code> </p>
        </td>
        <td><p>Import phased map list from polymapR</p></td>
      </tr><tr><td>
          <p><code><a href="loglike_hmm.html">loglike_hmm()</a></code> </p>
        </td>
        <td><p>Multipoint log-likelihood computation</p></td>
      </tr><tr><td>
          <p><code><a href="make_mat_mappoly.html">make_mat_mappoly()</a></code> </p>
        </td>
        <td><p>Subset recombination fraction matrices</p></td>
      </tr><tr><td>
          <p><code><a href="make_pairs_mappoly.html">make_pairs_mappoly()</a></code> </p>
        </td>
        <td><p>Subset pairwise recombination fractions</p></td>
      </tr><tr><td>
          <p><code><a href="make_seq_mappoly.html">make_seq_mappoly()</a></code> <code><a href="make_seq_mappoly.html">print(<i>&lt;mappoly.sequence&gt;</i>)</a></code> <code><a href="make_seq_mappoly.html">plot(<i>&lt;mappoly.sequence&gt;</i>)</a></code> </p>
        </td>
        <td><p>Create a sequence of markers</p></td>
      </tr><tr><td>
          <p><code><a href="maps.hexafake.html">maps.hexafake</a></code> </p>
        </td>
        <td><p>Resulting maps from <code><a href="https://rdrr.io/pkg/mappoly/man/hexafake.html" class="external-link">hexafake</a></code></p></td>
      </tr><tr><td>
          <p><code><a href="mds_mappoly.html">mds_mappoly()</a></code> <code><a href="mds_mappoly.html">print(<i>&lt;mappoly.pcmap&gt;</i>)</a></code> <code><a href="mds_mappoly.html">print(<i>&lt;mappoly.pcmap3d&gt;</i>)</a></code> </p>
        </td>
        <td><p>Estimates loci position using Multidimensional Scaling</p></td>
      </tr><tr><td>
          <p><code><a href="merge_datasets.html">merge_datasets()</a></code> </p>
        </td>
        <td><p>Merge datasets</p></td>
      </tr><tr><td>
          <p><code><a href="merge_maps.html">merge_maps()</a></code> </p>
        </td>
        <td><p>Merge two maps</p></td>
      </tr><tr><td>
          <p><code><a href="plot.mappoly.homoprob.html">plot(<i>&lt;mappoly.homoprob&gt;</i>)</a></code> </p>
        </td>
        <td><p>Plots mappoly.homoprob</p></td>
      </tr><tr><td>
          <p><code><a href="plot.mappoly.prefpair.profiles.html">plot(<i>&lt;mappoly.prefpair.profiles&gt;</i>)</a></code> </p>
        </td>
        <td><p>Plots mappoly.prefpair.profiles</p></td>
      </tr><tr><td>
          <p><code><a href="plot_GIC.html">plot_GIC()</a></code> </p>
        </td>
        <td><p>Genotypic information content</p></td>
      </tr><tr><td>
          <p><code><a href="plot_genome_vs_map.html">plot_genome_vs_map()</a></code> </p>
        </td>
        <td><p>Physical versus genetic distance</p></td>
      </tr><tr><td>
          <p><code><a href="plot_map_list.html">plot_map_list()</a></code> </p>
        </td>
        <td><p>Plot a genetic map</p></td>
      </tr><tr><td>
          <p><code><a href="plot_mappoly.map2.html">plot_mappoly.map2()</a></code> </p>
        </td>
        <td><p>Plot object mappoly.map2</p></td>
      </tr><tr><td>
          <p><code><a href="plot_mrk_info.html">plot_mrk_info()</a></code> </p>
        </td>
        <td><p>Plot marker information</p></td>
      </tr><tr><td>
          <p><code><a href="plot_progeny_dosage_change.html">plot_progeny_dosage_change()</a></code> </p>
        </td>
        <td><p>Look at genotypes that were imputed or changed by the HMM chain given a level of global genotypic error</p></td>
      </tr><tr><td>
          <p><code><a href="print_mrk.html">print_mrk()</a></code> </p>
        </td>
        <td><p>Summary of a set of markers</p></td>
      </tr><tr><td>
          <p><code><a href="read_fitpoly.html">read_fitpoly()</a></code> </p>
        </td>
        <td><p>Data Input in fitPoly format</p></td>
      </tr><tr><td>
          <p><code><a href="read_geno.html">read_geno()</a></code> <code><a href="read_geno.html">print(<i>&lt;mappoly.data&gt;</i>)</a></code> <code><a href="read_geno.html">plot(<i>&lt;mappoly.data&gt;</i>)</a></code> </p>
        </td>
        <td><p>Data Input</p></td>
      </tr><tr><td>
          <p><code><a href="read_geno_csv.html">read_geno_csv()</a></code> </p>
        </td>
        <td><p>Data Input in CSV format</p></td>
      </tr><tr><td>
          <p><code><a href="read_geno_prob.html">read_geno_prob()</a></code> </p>
        </td>
        <td><p>Data Input</p></td>
      </tr><tr><td>
          <p><code><a href="read_vcf.html">read_vcf()</a></code> </p>
        </td>
        <td><p>Data Input VCF</p></td>
      </tr><tr><td>
          <p><code><a href="reest_rf.html">reest_rf()</a></code> </p>
        </td>
        <td><p>Re-estimate the recombination fractions in a genetic map</p></td>
      </tr><tr><td>
          <p><code><a href="rev_map.html">rev_map()</a></code> </p>
        </td>
        <td><p>Reverse map</p></td>
      </tr><tr><td>
          <p><code><a href="rf_list_to_matrix.html">rf_list_to_matrix()</a></code> <code><a href="rf_list_to_matrix.html">print(<i>&lt;mappoly.rf.matrix&gt;</i>)</a></code> <code><a href="rf_list_to_matrix.html">plot(<i>&lt;mappoly.rf.matrix&gt;</i>)</a></code> </p>
        </td>
        <td><p>Recombination fraction list to matrix</p></td>
      </tr><tr><td>
          <p><code><a href="rf_snp_filter.html">rf_snp_filter()</a></code> </p>
        </td>
        <td><p>Remove markers that do not meet a LOD criteria</p></td>
      </tr><tr><td>
          <p><code><a href="segreg_poly.html">segreg_poly()</a></code> </p>
        </td>
        <td><p>Polysomic segregation frequency</p></td>
      </tr><tr><td>
          <p><code><a href="sim_homologous.html">sim_homologous()</a></code> </p>
        </td>
        <td><p>Simulate homology groups</p></td>
      </tr><tr><td>
          <p><code><a href="solcap.dose.map.html">solcap.dose.map</a></code> </p>
        </td>
        <td><p>Resulting maps from <code><a href="https://rdrr.io/pkg/mappoly/man/tetra.solcap.html" class="external-link">tetra.solcap</a></code></p></td>
      </tr><tr><td>
          <p><code><a href="solcap.err.map.html">solcap.err.map</a></code> </p>
        </td>
        <td><p>Resulting maps from <code><a href="https://rdrr.io/pkg/mappoly/man/tetra.solcap.html" class="external-link">tetra.solcap</a></code></p></td>
      </tr><tr><td>
          <p><code><a href="solcap.mds.map.html">solcap.mds.map</a></code> </p>
        </td>
        <td><p>Resulting maps from <code><a href="https://rdrr.io/pkg/mappoly/man/tetra.solcap.html" class="external-link">tetra.solcap</a></code></p></td>
      </tr><tr><td>
          <p><code><a href="solcap.prior.map.html">solcap.prior.map</a></code> </p>
        </td>
        <td><p>Resulting maps from <code><a href="https://rdrr.io/pkg/mappoly/man/tetra.solcap.geno.dist.html" class="external-link">tetra.solcap.geno.dist</a></code></p></td>
      </tr><tr><td>
          <p><code><a href="split_and_rephase.html">split_and_rephase()</a></code> </p>
        </td>
        <td><p>Divides map in sub-maps and re-phase them</p></td>
      </tr><tr><td>
          <p><code><a href="summary_maps.html">summary_maps()</a></code> </p>
        </td>
        <td><p>Summary maps</p></td>
      </tr><tr><td>
          <p><code><a href="tetra.solcap.html">tetra.solcap</a></code> </p>
        </td>
        <td><p>Autotetraploid potato dataset.</p></td>
      </tr><tr><td>
          <p><code><a href="tetra.solcap.geno.dist.html">tetra.solcap.geno.dist</a></code> </p>
        </td>
        <td><p>Autotetraploid potato dataset with genotype probabilities.</p></td>
      </tr><tr><td>
          <p><code><a href="update_framework_map.html">update_framework_map()</a></code> </p>
        </td>
        <td><p>Add markers that are informative in both parents using HMM approach and evaluating difference 
in LOD and gap size</p></td>
      </tr><tr><td>
          <p><code><a href="update_map.html">update_map()</a></code> </p>
        </td>
        <td><p>Update map</p></td>
      </tr><tr><td>
          <p><code><a href="update_missing.html">update_missing()</a></code> </p>
        </td>
        <td><p>Update missing information</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marcelo Mollinari, Gabriel Gesteira, Cristiane Taniguti, Jeekin Lau.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

