<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Merge datasets — merge_datasets • mappoly</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Merge datasets — merge_datasets"><meta property="og:description" content="This function merges two datasets of class mappoly.data. This can be useful
when individuals of a population were genotyped using two or more techniques
and have datasets in different files or formats. Please notice that the datasets 
should contain the same number of individuals and they must be represented identically 
in both datasets  (e.g. Ind_1 in both datasets, not Ind_1
in one dataset and ind_1 or Ind.1 in the other)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mappoly</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/mappoly_startguide.html">Introduction to MAPpoly</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mmollina/MAPpoly/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Merge datasets</h1>
    <small class="dont-index">Source: <a href="https://github.com/mmollina/MAPpoly/blob/HEAD/R/utils.R" class="external-link"><code>R/utils.R</code></a></small>
    <div class="hidden name"><code>merge_datasets.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function merges two datasets of class <code>mappoly.data</code>. This can be useful
when individuals of a population were genotyped using two or more techniques
and have datasets in different files or formats. Please notice that the datasets 
should contain the same number of individuals and they must be represented identically 
in both datasets  (e.g. <code>Ind_1</code> in both datasets, not <code>Ind_1</code>
in one dataset and <code>ind_1</code> or <code>Ind.1</code> in the other).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">merge_datasets</span><span class="op">(</span>dat.1 <span class="op">=</span> <span class="cn">NULL</span>, dat.2 <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>dat.1</dt>
<dd><p>the first dataset of class <code>mappoly.data</code> to be merged</p></dd>


<dt>dat.2</dt>
<dd><p>the second dataset of class <code>mappoly.data</code> to be merged (default = NULL);
if <code>dat.2 = NULL</code>, the function returns <code>dat.1</code> only</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An object of class <code>mappoly.data</code> which contains all markers
from both datasets. It will be a list with the following components:</p>
<dl><dt>ploidy</dt>
<dd><p>ploidy level</p></dd>

    <dt>n.ind</dt>
<dd><p>number individuals</p></dd>

    <dt>n.mrk</dt>
<dd><p>total number of markers</p></dd>

    <dt>ind.names</dt>
<dd><p>the names of the individuals</p></dd>

    <dt>mrk.names</dt>
<dd><p>the names of the markers</p></dd>

    <dt>dosage.p1</dt>
<dd><p>a vector containing the dosage in
      parent P for all <code>n.mrk</code> markers</p></dd>

    <dt>dosage.p2</dt>
<dd><p>a vector containing the dosage in
      parent Q for all <code>n.mrk</code> markers</p></dd>

    <dt>chrom</dt>
<dd><p>a vector indicating which sequence each marker
      belongs. Zero indicates that the marker was not assigned to any
      sequence</p></dd>

    <dt>genome.pos</dt>
<dd><p>Physical position of the markers into the
      sequence</p></dd>

    <dt>seq.ref</dt>
<dd><p>if one or both datasets originated from read_vcf, it keeps
reference alleles from sequencing platform, otherwise is NULL</p></dd>

    <dt>seq.alt</dt>
<dd><p>if one or both datasets originated from read_vcf, it keeps
alternative alleles from sequencing platform, otherwise is NULL</p></dd>

    <dt>all.mrk.depth</dt>
<dd><p>if one or both datasets originated from read_vcf, it keeps
marker read depths from sequencing, otherwise is NULL</p></dd>

    <dt>prob.thres</dt>
<dd><p>(unused field)</p></dd>

    <dt>geno.dose</dt>
<dd><p>a matrix containing the dosage for each markers (rows) 
      for each individual (columns). Missing data are represented by 
      <code>ploidy_level + 1</code></p></dd>

    <dt>geno</dt>
<dd><p>if both datasets contain genotype distribution information,
the final object will contain 'geno'. This is set to NULL otherwise</p></dd>

    <dt>nphen</dt>
<dd><p>(0)</p></dd>

    <dt>phen</dt>
<dd><p>(NULL)</p></dd>

    <dt>chisq.pval</dt>
<dd><p>a vector containing p-values related to the chi-squared 
    test of Mendelian segregation performed for all markers in both datasets</p></dd>

    <dt>kept</dt>
<dd><p>if elim.redundant = TRUE when reading any dataset, holds all non-redundant markers</p></dd>

    <dt>elim.correspondence</dt>
<dd><p>if elim.redundant = TRUE when reading any dataset,
holds all non-redundant markers and its equivalence to the redundant ones</p></dd>

</dl></div>
    <div id="references">
    <h2>References</h2>
    <p>Mollinari, M., and Garcia, A.  A. F. (2019) Linkage
    analysis and haplotype phasing in experimental autopolyploid
    populations with high ploidy level using hidden Markov
    models, _G3: Genes, Genomes, Genetics_. 
    doi: <a href="https://doi.org/10.1534/g3.119.400378" class="external-link">10.1534/g3.119.400378</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Gabriel Gesteira, <a href="mailto:gdesiqu@ncsu.edu">gdesiqu@ncsu.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="co">## Loading a subset of SNPs from chromosomes 3 and 12 of sweetpotato dataset </span></span></span>
<span class="r-in"><span><span class="co">## (SNPs anchored to Ipomoea trifida genome)</span></span></span>
<span class="r-in"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></span>
<span class="r-in"><span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">12</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Loading chromosome"</span>, <span class="va">i</span>, <span class="st">"...\n"</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">tempfl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"ch"</span>, <span class="va">i</span><span class="op">)</span>, fileext <span class="op">=</span> <span class="st">".vcf.gz"</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">x</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/mmollina/MAPpoly_vignettes/raw/master/data/sweet_sample_ch"</span></span></span>
<span class="r-in"><span>    <span class="va">address</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">i</span>, <span class="st">".vcf.gz"</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="va">address</span>, destfile <span class="op">=</span> <span class="va">tempfl</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">dattemp</span> <span class="op">&lt;-</span> <span class="fu"><a href="read_vcf.html">read_vcf</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">tempfl</span>, parent.1 <span class="op">=</span> <span class="st">"PARENT1"</span>, parent.2 <span class="op">=</span> <span class="st">"PARENT2"</span>,</span></span>
<span class="r-in"><span>                        ploidy <span class="op">=</span> <span class="fl">6</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">merge_datasets</span><span class="op">(</span><span class="va">dat</span>, <span class="va">dattemp</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Loading chromosome 3 ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Registered S3 method overwritten by 'vegan':</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   method     from      </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   rev.hclust dendextend</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Loading chromosome 12 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">dat</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is an object of class 'mappoly.data'</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Ploidy level:                            6 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. individuals:                         315 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. markers:                             310 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Missing data:                            6.19%</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Redundant markers:                       1.27%</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     This dataset contains chromosome information.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ----------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. of markers per dosage combination in both parents:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     P1 P2 freq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      0  1   91</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      0  2   14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      0  3    7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      0  4    2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1  0   41</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1  1   26</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1  2   23</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1  3    9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1  4    3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1  5    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2  0    9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2  1   17</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2  2   12</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2  3   13</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2  4    5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      3  0    5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      3  1    7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      3  2   14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      3  3    5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      4  0    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      4  1    3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      4  2    2</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="merge_datasets-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marcelo Mollinari, Gabriel Gesteira.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

