<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Data Input in fitPoly format — read_fitpoly • mappoly</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Data Input in fitPoly format — read_fitpoly"><meta property="og:description" content="Reads an external data file generated as output of saveMarkerModels. 
This function creates an object of class mappoly.data."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mappoly</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/mappoly_startguide.html">Introduction to MAPpoly</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mmollina/MAPpoly/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Data Input in fitPoly format</h1>
    <small class="dont-index">Source: <a href="https://github.com/mmollina/MAPpoly/blob/HEAD/R/read_fitpoly.R" class="external-link"><code>R/read_fitpoly.R</code></a></small>
    <div class="hidden name"><code>read_fitpoly.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Reads an external data file generated as output of <code><a href="https://rdrr.io/pkg/fitPoly/man/saveMarkerModels.html" class="external-link">saveMarkerModels</a></code>. 
This function creates an object of class <code>mappoly.data</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">read_fitpoly</span><span class="op">(</span></span>
<span>  <span class="va">file.in</span>,</span>
<span>  <span class="va">ploidy</span>,</span>
<span>  <span class="va">parent1</span>,</span>
<span>  <span class="va">parent2</span>,</span>
<span>  offspring <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filter.non.conforming <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  elim.redundant <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  parent.geno <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"joint"</span>, <span class="st">"max"</span><span class="op">)</span>,</span>
<span>  thresh.parent.geno <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  prob.thres <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  file.type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"table"</span>, <span class="st">"csv"</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>file.in</dt>
<dd><p>a character string with the name of (or full path to) the input file</p></dd>


<dt>ploidy</dt>
<dd><p>the ploidy level</p></dd>


<dt>parent1</dt>
<dd><p>a character string containing the name (or pattern of genotype IDs) of parent 1</p></dd>


<dt>parent2</dt>
<dd><p>a character string containing the name (or pattern of genotype IDs) of parent 2</p></dd>


<dt>offspring</dt>
<dd><p>a character string containing the name (or pattern of genotype IDs) of the offspring 
individuals. If <code>NULL</code> (default) it considers all individuals as offsprings, except 
<code>parent1</code> and <code>parent2</code>.</p></dd>


<dt>filter.non.conforming</dt>
<dd><p>if <code>TRUE</code> (default) converts data points with unexpected 
genotypes (i.e. no double reduction) to 'NA'. See function <code><a href="segreg_poly.html">segreg_poly</a></code> 
for information on expected classes and their respective frequencies.</p></dd>


<dt>elim.redundant</dt>
<dd><p>logical. If <code>TRUE</code> (default), removes redundant markers
during map construction, keeping them annotated to in order to include them in the final map.</p></dd>


<dt>parent.geno</dt>
<dd><p>indicates whether to use the joint probability <code>'joint'</code> (default) or the 
maximum probability of multiple replicates (if available) to assign dosage to parents. 
If there is one observation per parent, both options will yield the same results.</p></dd>


<dt>thresh.parent.geno</dt>
<dd><p>threshold probability to assign a dosage to parents. If the probability 
is smaller than <code>thresh.parent.geno</code>, the marker is discarded.</p></dd>


<dt>prob.thres</dt>
<dd><p>threshold probability to assign a dosage to offspring. If the probability 
is smaller than <code>prob.thres</code>, the data point is converted to 'NA'.</p></dd>


<dt>file.type</dt>
<dd><p>indicates whether the characters in the input file are separated by 
'white spaces' ("table") or by commas ("csv").</p></dd>


<dt>verbose</dt>
<dd><p>if <code>TRUE</code> (default), the current progress is shown; if
<code>FALSE</code>, no output is produced</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An object of class <code>mappoly.data</code> which contains a
    list with the following components:</p>
<dl><dt>ploidy</dt>
<dd><p>ploidy level</p></dd>

    <dt>n.ind</dt>
<dd><p>number individuals</p></dd>

    <dt>n.mrk</dt>
<dd><p>total number of markers</p></dd>

    <dt>ind.names</dt>
<dd><p>the names of the individuals</p></dd>

    <dt>mrk.names</dt>
<dd><p>the names of the markers</p></dd>

    <dt>dosage.p1</dt>
<dd><p>a vector containing the dosage in
      parent P for all <code>n.mrk</code> markers</p></dd>

    <dt>dosage.p2</dt>
<dd><p>a vector containing the dosage in
      parent Q for all <code>n.mrk</code> markers</p></dd>

    <dt>chrom</dt>
<dd><p>a vector indicating which sequence each marker
      belongs. Zero indicates that the marker was not assigned to any
      sequence</p></dd>

    <dt>genome.pos</dt>
<dd><p>Physical position of the markers into the
      sequence</p></dd>

    <dt>seq.ref</dt>
<dd><p>NULL (unused in this type of data)</p></dd>

    <dt>seq.alt</dt>
<dd><p>NULL (unused in this type of data)</p></dd>

    <dt>all.mrk.depth</dt>
<dd><p>NULL (unused in this type of data)</p></dd>

    <dt>geno.dose</dt>
<dd><p>a matrix containing the dosage for each markers (rows) 
      for each individual (columns). Missing data are represented by 
      <code>ploidy_level + 1</code></p></dd>

    <dt>n.phen</dt>
<dd><p>number of phenotypic traits</p></dd>

    <dt>phen</dt>
<dd><p>a matrix containing the phenotypic data. The rows
                correspond to the traits and the columns correspond
                to the individuals</p></dd>

    <dt>kept</dt>
<dd><p>if elim.redundant = TRUE, holds all non-redundant markers</p></dd>

    <dt>elim.correspondence</dt>
<dd><p>if elim.redundant = TRUE, holds all non-redundant markers and
its equivalence to the redundant ones</p></dd>

</dl></div>
    <div id="references">
    <h2>References</h2>
    <p>Voorrips, R.E., Gort, G. &amp; Vosman, B. (2011) Genotype calling 
    in tetraploid species from bi-allelic marker data using mixture 
    models. _BMC Bioinformatics_.
    <a href="https://doi.org/10.1186/1471-2105-12-172" class="external-link">doi:10.1186/1471-2105-12-172</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Marcelo Mollinari, <a href="mailto:mmollin@ncsu.edu">mmollin@ncsu.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="co">#### Tetraploid Example</span></span></span>
<span class="r-in"><span><span class="va">ft</span> <span class="op">&lt;-</span> <span class="st">"https://raw.githubusercontent.com/mmollina/MAPpoly_vignettes/master/data/fitpoly.dat"</span></span></span>
<span class="r-in"><span><span class="va">tempfl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">ft</span>, destfile <span class="op">=</span> <span class="va">tempfl</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fitpoly.dat</span> <span class="op">&lt;-</span> <span class="fu">read_fitpoly</span><span class="op">(</span>file.in <span class="op">=</span> <span class="va">tempfl</span>, ploidy <span class="op">=</span> <span class="fl">4</span>, </span></span>
<span class="r-in"><span>                            parent1 <span class="op">=</span> <span class="st">"P1"</span>, parent2 <span class="op">=</span> <span class="st">"P2"</span>, </span></span>
<span class="r-in"><span>                            verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reading the following data:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Ploidy level: 4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. individuals:  150</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. markers:  145</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. informative markers:  138 (95.2%)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Done with reading.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Filtering non-conforming markers.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Performing chi-square test.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Done.</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">fitpoly.dat</span>, detailed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is an object of class 'mappoly.data'</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Ploidy level:                            4 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. individuals:                         150 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. markers:                             138 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Prob. threshold to declare missing:      0.95 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Missing data:                            3.89%</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. markers per chromosome: not available</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ----------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. of markers per dosage combination in both parents:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     P1 P2 freq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      0  1   18</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      0  2   24</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1  0   18</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1  1   19</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1  2   10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1  3    3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2  0   13</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2  1    6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2  2   15</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2  3    2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      3  1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      3  2    5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      3  3    2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      3  4    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      4  2    1</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fitpoly.dat</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="read_fitpoly-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="fu"><a href="plot_mrk_info.html">plot_mrk_info</a></span><span class="op">(</span><span class="va">fitpoly.dat</span>, <span class="fl">37</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="read_fitpoly-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marcelo Mollinari, Gabriel Gesteira, Cristiane Taniguti.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

