<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Data Input VCF — read_vcf • mappoly</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Data Input VCF — read_vcf"><meta property="og:description" content="Reads an external VCF file and creates an object of class mappoly.data"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mappoly</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/mappoly_startguide.html">Introduction to MAPpoly</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mmollina/MAPpoly/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Data Input VCF</h1>
    <small class="dont-index">Source: <a href="https://github.com/mmollina/MAPpoly/blob/HEAD/R/read_mappoly_vcf.R" class="external-link"><code>R/read_mappoly_vcf.R</code></a></small>
    <div class="hidden name"><code>read_vcf.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Reads an external VCF file and creates an object of class <code>mappoly.data</code></p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">read_vcf</span><span class="op">(</span></span>
<span>  <span class="va">file.in</span>,</span>
<span>  <span class="va">parent.1</span>,</span>
<span>  <span class="va">parent.2</span>,</span>
<span>  ploidy <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  filter.non.conforming <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  thresh.line <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  min.gt.depth <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  min.av.depth <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  max.missing <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  elim.redundant <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  read.geno.prob <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  prob.thres <span class="op">=</span> <span class="fl">0.95</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>file.in</dt>
<dd><p>a character string with the name of (or full path to) the 
input file which contains the data (VCF format)</p></dd>


<dt>parent.1</dt>
<dd><p>a character string containing the name of parent 1</p></dd>


<dt>parent.2</dt>
<dd><p>a character string containing the name of parent 2</p></dd>


<dt>ploidy</dt>
<dd><p>the species ploidy (optional, it will be automatically detected)</p></dd>


<dt>filter.non.conforming</dt>
<dd><p>if <code>TRUE</code> (default) converts data points with unexpected 
genotypes (i.e. no double reduction) to 'NA'. See function <code><a href="segreg_poly.html">segreg_poly</a></code> 
for information on expected classes and their respective frequencies.</p></dd>


<dt>thresh.line</dt>
<dd><p>threshold used for p-values on segregation test (default = 0.05)</p></dd>


<dt>min.gt.depth</dt>
<dd><p>minimum genotype depth to keep information. 
If the genotype depth is below <code>min.gt.depth</code>,
it will be replaced with NA (default = 0)</p></dd>


<dt>min.av.depth</dt>
<dd><p>minimum average depth to keep markers (default = 0)</p></dd>


<dt>max.missing</dt>
<dd><p>maximum proportion of missing data to keep markers (range = 0-1; default = 1)</p></dd>


<dt>elim.redundant</dt>
<dd><p>logical. If <code>TRUE</code> (default), removes redundant markers
during map construction, keeping them annotated to export to the final map.</p></dd>


<dt>verbose</dt>
<dd><p>if <code>TRUE</code> (default), the current progress is shown; if
<code>FALSE</code>, no output is produced</p></dd>


<dt>read.geno.prob</dt>
<dd><p>If genotypic probabilities are available (PL field),
generates a probability-based dataframe (default = <code>FALSE</code>).</p></dd>


<dt>prob.thres</dt>
<dd><p>probability threshold to associate a marker call to a 
dosage. Markers with maximum genotype probability smaller than <code>prob.thres</code> 
are considered as missing data for the dosage calling purposes (default = 0.95)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An object of class <code>mappoly.data</code> which contains a
    list with the following components:</p>
<dl><dt>ploidy</dt>
<dd><p>ploidy level</p></dd>

    <dt>n.ind</dt>
<dd><p>number individuals</p></dd>

    <dt>n.mrk</dt>
<dd><p>total number of markers</p></dd>

    <dt>ind.names</dt>
<dd><p>the names of the individuals</p></dd>

    <dt>mrk.names</dt>
<dd><p>the names of the markers</p></dd>

    <dt>dosage.p1</dt>
<dd><p>a vector containing the dosage in
      parent P for all <code>n.mrk</code> markers</p></dd>

    <dt>dosage.p2</dt>
<dd><p>a vector containing the dosage in
      parent Q for all <code>n.mrk</code> markers</p></dd>

    <dt>chrom</dt>
<dd><p>a vector indicating which sequence each marker
      belongs. Zero indicates that the marker was not assigned to any
      sequence</p></dd>

    <dt>genome.pos</dt>
<dd><p>Physical position of the markers into the
      sequence</p></dd>

    <dt>seq.ref</dt>
<dd><p>Reference base used for each marker (i.e. A, T, C, G)</p></dd>

    <dt>seq.alt</dt>
<dd><p>Alternative base used for each marker (i.e. A, T, C, G)</p></dd>

    <dt>prob.thres</dt>
<dd><p>(unused field)</p></dd>

    <dt>geno.dose</dt>
<dd><p>a matrix containing the dosage for each markers (rows) 
      for each individual (columns). Missing data are represented by 
      <code>ploidy_level + 1</code></p></dd>

    <dt>geno</dt>
<dd><p>a dataframe containing all genotypic probabilities columns for each
      marker and individual combination (rows). Missing data are represented by 
      <code>ploidy_level + 1</code></p></dd>

    <dt>nphen</dt>
<dd><p>(unused field)</p></dd>

    <dt>phen</dt>
<dd><p>(unused field)</p></dd>

    <dt>all.mrk.depth</dt>
<dd><p>DP information for all markers on VCF file</p></dd>

    <dt>chisq.pval</dt>
<dd><p>a vector containing p-values related to the chi-squared 
    test of Mendelian segregation performed for all markers</p></dd>

    <dt>kept</dt>
<dd><p>if elim.redundant = TRUE, holds all non-redundant markers</p></dd>

    <dt>elim.correspondence</dt>
<dd><p>if elim.redundant = TRUE, holds all non-redundant markers and
its equivalence to the redundant ones</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>This function can handle .vcf files versions 4.0 or higher. The ploidy 
can be automatically detected, but it is highly recommended that you 
inform it to check for mismatches. All individual and marker names 
will be kept as they are in the .vcf file.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Mollinari M., Olukolu B. A.,  Pereira G. da S., 
    Khan A., Gemenet D., Yencho G. C., Zeng Z-B. (2020), 
    Unraveling the Hexaploid Sweetpotato Inheritance 
    Using Ultra-Dense Multilocus Mapping, 
    _G3: Genes, Genomes, Genetics_. 
    <a href="https://doi.org/10.1534/g3.119.400620" class="external-link">doi:10.1534/g3.119.400620</a></p>    
<p>Mollinari, M., and Garcia, A.  A. F. (2019) Linkage
    analysis and haplotype phasing in experimental autopolyploid
    populations with high ploidy level using hidden Markov
    models, _G3: Genes, Genomes, Genetics_. 
    <a href="https://doi.org/10.1534/g3.119.400378" class="external-link">doi:10.1534/g3.119.400378</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Gabriel Gesteira, <a href="mailto:gdesiqu@ncsu.edu">gdesiqu@ncsu.edu</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="co">## Hexaploid sweetpotato: Subset of chromosome 3</span></span></span>
<span class="r-in"><span><span class="va">fl</span> <span class="op">=</span> <span class="st">"https://github.com/mmollina/MAPpoly_vignettes/raw/master/data/sweet_sample_ch3.vcf.gz"</span></span></span>
<span class="r-in"><span><span class="va">tempfl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">'chr3_'</span>, fileext <span class="op">=</span> <span class="st">'.vcf.gz'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">fl</span>, destfile <span class="op">=</span> <span class="va">tempfl</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dat.dose.vcf</span> <span class="op">=</span> <span class="fu">read_vcf</span><span class="op">(</span>file <span class="op">=</span> <span class="va">tempfl</span>, parent.1 <span class="op">=</span> <span class="st">"PARENT1"</span>, parent.2 <span class="op">=</span> <span class="st">"PARENT2"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reading data...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Scanning file to determine attributes.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> File attributes:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   meta lines: 8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   header_line: 9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   variant count: 200</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   column count: 326</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Meta line 8 read in.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> All meta lines processed.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gt matrix initialized.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Character matrix gt created.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Character matrix gt rows: 200</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Character matrix gt cols: 326</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   skip: 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   nrows: 200</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   row_num: 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Processed variant: 200</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> All variants processed</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Processing genotypes...Done!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Selected ploidy: 6 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Read the following data:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Ploidy level: 6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. individuals:  315</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. markers:  167</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. informative markers:  167 (100%)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     This dataset contains chromosome information.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Done with reading.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Filtering non-conforming markers.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Performing chi-square test.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Done.</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">dat.dose.vcf</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This is an object of class 'mappoly.data'</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Ploidy level:                            6 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. individuals:                         315 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. markers:                             164 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Missing data under  prob. threshold: 6.01%</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     Redundant markers:                       1.8%</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. markers per chromosome: not available</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     ----------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     No. of markers per dosage combination in both parents:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     P1 P2 freq</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      0  1   50</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      0  2    8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      0  3    4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      0  4    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1  0   26</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1  1   12</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1  2    9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1  3    4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      1  4    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2  0    3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2  1   10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2  2    7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2  3    7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      2  4    4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      3  0    3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      3  1    3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      3  2    7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      3  3    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      4  0    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      4  1    2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      4  2    1</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">dat.dose.vcf</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="read_vcf-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marcelo Mollinari, Gabriel Gesteira.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

